{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment Project: hg-002 controldb SV check\n",
    "\n",
    "Follow-up from enrichment (Layman & NTD) discussion with Nikhil (GGC) regarding large average # of SVs per patient. HG-002 (GM24385) is filtered then clustered.\n",
    "\n",
    "An addition to this script is to test compilation of Enrichment Workflow Steps 1 & 2 in a util.py"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Enrichment Single Clustering (Filter then cluster)\n",
    "Troubleshooting clustering code in cohort enrichment project."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load packages\n",
    "from itertools import groupby\n",
    "from matplotlib_venn import venn2, venn2_circles\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "from enrichment_util import *\n",
    "\n",
    "import copy\n",
    "import re \n",
    "import os\n",
    "import pathlib\n",
    "working_dir = pathlib.Path().absolute()\n",
    "os.chdir(working_dir)\n",
    "\n",
    "\n",
    "%matplotlib inline\n",
    "pd.options.mode.chained_assignment = None  # default='warn'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standardized headers for both bed files\n",
    "# headers: chrom\tchromStart\tchromEnd\tGene\tIndex\tstrand\tchromStart2\tchromEnd\tRGB\n",
    "bed_headers = [\"chr\",\"RefStartPos\",\"RefEndPos\",\"Gene\",\"Index\",\"Strand\",\"RefStartPos2\",\"RefEndPos2\",\"RGB\"]\n",
    "\n",
    "# Load hg38.knownCanonical.bed\n",
    "bed_df = pd.read_csv(r\"Input\\ref_genes\\hg38.knownCanonical.mapped_KH.bed\", dtype=object, index_col=False, sep=\"\\t\")\n",
    "bed_df = bed_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "\n",
    "# Load hg38_gaps.bed\n",
    "gaps_df = pd.read_csv(r\"Input\\ref_genes\\hg38_gaps.bed\", dtype=object,  index_col=False, sep=\"\\t\", names=bed_headers)\n",
    "gaps_df = gaps_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "\n",
    "# Load hg38 pseudogene list.bed\n",
    "pseudo_df = pd.read_csv(r\"Input\\ref_genes\\hg38_EncodeGencodePseudoGeneV41.bed\",dtype=object,  index_col=False, sep=\"\\t\")\n",
    "pseudo_df = pseudo_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "\n",
    "# set redstartstop as float for downstream filtering functions\n",
    "for df in [bed_df, gaps_df, pseudo_df]:\n",
    "    df = df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "WindowsPath('c:/Users/sshukor/OneDrive - Bionano Genomics/Documents/Enrichment Project/clustering_optimization_enrichment')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "working_dir"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get smap column header and table rows, save as *_minimal.smap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = \"Input\\controlsdb\\GM24385\\exp_refineFinal1_merged_filter_inversions.smap\"\n",
    "out_path = \"Output\\output_hg002_clustering_04Jul2023\\GM24385_minimal.smap\"\n",
    "# get_smap_table(in_path, out_path)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "load sv table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in hg002 .smap\n",
    "# load csv as DF\n",
    "hg002_path = \"Output\\output_hg002_clustering_04Jul2023\\GM24385_minimal.smap\"\n",
    "df = pd.read_csv(hg002_path, sep='\\t', dtype=object,  index_col=False)\n",
    "df = df.assign(Sample_ID = \"GM24385\")\n",
    "df.rename(columns={\"RefcontigID1\":\"chr\",\"SmapEntryID\":\"SmapID\"}, inplace=True)\n",
    "df = df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'SVsize':'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SmapID</th>\n",
       "      <th>QryContigID</th>\n",
       "      <th>chr</th>\n",
       "      <th>RefcontigID2</th>\n",
       "      <th>QryStartPos</th>\n",
       "      <th>QryEndPos</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>Type</th>\n",
       "      <th>...</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>GenotypeGroup</th>\n",
       "      <th>RawConfidence</th>\n",
       "      <th>RawConfidenceLeft</th>\n",
       "      <th>RawConfidenceRight</th>\n",
       "      <th>RawConfidenceCenter</th>\n",
       "      <th>SVsize</th>\n",
       "      <th>SVfreq</th>\n",
       "      <th>orientation</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>3121</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36313.3</td>\n",
       "      <td>60678.2</td>\n",
       "      <td>365535.0</td>\n",
       "      <td>387230.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>insertion</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4.22</td>\n",
       "      <td>4.22</td>\n",
       "      <td>25.61</td>\n",
       "      <td>34.43</td>\n",
       "      <td>2669.9</td>\n",
       "      <td>0.070</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>681</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2175119.8</td>\n",
       "      <td>2230505.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>617119.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3.00</td>\n",
       "      <td>198.52</td>\n",
       "      <td>3669.43</td>\n",
       "      <td>561733.6</td>\n",
       "      <td>0.620</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>681</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2119878.4</td>\n",
       "      <td>2128379.7</td>\n",
       "      <td>663861.0</td>\n",
       "      <td>672078.0</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>insertion</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.19</td>\n",
       "      <td>9.24</td>\n",
       "      <td>148.70</td>\n",
       "      <td>1.19</td>\n",
       "      <td>284.4</td>\n",
       "      <td>0.443</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>681</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2079978.7</td>\n",
       "      <td>2081775.4</td>\n",
       "      <td>710362.1</td>\n",
       "      <td>711828.9</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>insertion</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5.28</td>\n",
       "      <td>17.61</td>\n",
       "      <td>141.73</td>\n",
       "      <td>5.28</td>\n",
       "      <td>329.9</td>\n",
       "      <td>0.453</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>681</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1985981.4</td>\n",
       "      <td>2003404.8</td>\n",
       "      <td>788500.0</td>\n",
       "      <td>803475.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>insertion</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>27.74</td>\n",
       "      <td>27.74</td>\n",
       "      <td>131.69</td>\n",
       "      <td>85.80</td>\n",
       "      <td>2448.5</td>\n",
       "      <td>0.491</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  SmapID QryContigID  chr RefcontigID2 QryStartPos  QryEndPos  RefStartPos  \\\n",
       "0      4        3121    1            1     36313.3    60678.2     365535.0   \n",
       "1      5         681    1            1   2175119.8  2230505.2          0.0   \n",
       "2      6         681    1            1   2119878.4  2128379.7     663861.0   \n",
       "3      7         681    1            1   2079978.7  2081775.4     710362.1   \n",
       "4      8         681    1            1   1985981.4  2003404.8     788500.0   \n",
       "\n",
       "   RefEndPos Confidence       Type  ... Genotype GenotypeGroup RawConfidence  \\\n",
       "0   387230.0       0.99  insertion  ...        1             1          4.22   \n",
       "1   617119.0       0.99   deletion  ...        2             1          3.00   \n",
       "2   672078.0      -1.00  insertion  ...        1             2          1.19   \n",
       "3   711828.9      -1.00  insertion  ...        1             3          5.28   \n",
       "4   803475.0       0.99  insertion  ...        1             4         27.74   \n",
       "\n",
       "  RawConfidenceLeft RawConfidenceRight RawConfidenceCenter    SVsize SVfreq  \\\n",
       "0              4.22              25.61               34.43    2669.9  0.070   \n",
       "1              3.00             198.52             3669.43  561733.6  0.620   \n",
       "2              9.24             148.70                1.19     284.4  0.443   \n",
       "3             17.61             141.73                5.28     329.9  0.453   \n",
       "4             27.74             131.69               85.80    2448.5  0.491   \n",
       "\n",
       "  orientation Sample_ID  \n",
       "0         NA    GM24385  \n",
       "1         NA    GM24385  \n",
       "2         NA    GM24385  \n",
       "3         NA    GM24385  \n",
       "4         NA    GM24385  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "((17515, 28),\n",
       " insertion                 9959\n",
       " deletion                  5388\n",
       " deletion_tiny              961\n",
       " insertion_tiny             499\n",
       " inversion_partial          173\n",
       " inversion                  172\n",
       " translocation_intrachr     123\n",
       " translocation_interchr      66\n",
       " duplication                 59\n",
       " duplication_inverted        53\n",
       " inversion_paired            42\n",
       " trans_intrachr_repeat       13\n",
       " duplication_split            6\n",
       " inversion_repeat             1\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(df.head())\n",
    "df.shape, df['Type'].value_counts()\n",
    "# df.dtypes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "flatten inversions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['insertion', 'deletion', 'insertion_tiny', 'duplication_inverted',\n",
       "       'inversion', 'inversion_partial', 'duplication', 'deletion_tiny',\n",
       "       'inversion_paired', 'translocation_intrachr',\n",
       "       'translocation_interchr', 'inversion_repeat',\n",
       "       'trans_intrachr_repeat', 'duplication_split'], dtype=object)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.unique(df['Type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total: 17515\n",
      "inversions, inversion nbase, and inversion & inversion_partials: 388\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "((17321, 28),\n",
       " insertion                 9959\n",
       " deletion                  5388\n",
       " deletion_tiny              961\n",
       " insertion_tiny             499\n",
       " inversion                  172\n",
       " translocation_intrachr     123\n",
       " translocation_interchr      66\n",
       " duplication                 59\n",
       " duplication_inverted        53\n",
       " inversion_paired            21\n",
       " trans_intrachr_repeat       13\n",
       " duplication_split            6\n",
       " inversion_repeat             1\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inv_flat_df = flatten_df_inv(df)\n",
    "inv_flat_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_inv_flat_out.csv\", index=False)\n",
    "inv_flat_df.shape, inv_flat_df['Type'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "keep non-translocations and exclude gains/loss, *masked, *common, *tiny, *nbase and complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Filter rare SVs and output selectd columns\n",
    "# df_rareSV = filter_rare_svs(inv_flat_df, 1.0)\n",
    "# # # df_rareSV = filter_rare_svs(case_df, 1.0)\n",
    "# n_rare = len(df_rareSV) \n",
    "\n",
    "# exclude unwanted sv types\n",
    "df_excluded = exclude_types_svs(inv_flat_df)\n",
    "# df_excluded = exclude_types_svs(inv_flat_df)\n",
    "# df_excluded = exclude_types_svs(case_df)\n",
    "\n",
    "# filter for confidence intervals\n",
    "df_sig = filter_sig_svs(df_excluded)\n",
    "n_sig = len(df_sig)\n",
    "\n",
    "df_sig.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_filtered_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((6467, 28),\n",
       " insertion               4388\n",
       " deletion                1934\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sig.shape, df_sig['Type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SmapID</th>\n",
       "      <th>QryContigID</th>\n",
       "      <th>chr</th>\n",
       "      <th>RefcontigID2</th>\n",
       "      <th>QryStartPos</th>\n",
       "      <th>QryEndPos</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>Type</th>\n",
       "      <th>...</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>GenotypeGroup</th>\n",
       "      <th>RawConfidence</th>\n",
       "      <th>RawConfidenceLeft</th>\n",
       "      <th>RawConfidenceRight</th>\n",
       "      <th>RawConfidenceCenter</th>\n",
       "      <th>SVsize</th>\n",
       "      <th>SVfreq</th>\n",
       "      <th>orientation</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>707</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16633490.1</td>\n",
       "      <td>16633490.1</td>\n",
       "      <td>103711691.0</td>\n",
       "      <td>103734414.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>inversion_paired</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>16069.91</td>\n",
       "      <td>5.25</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>22723.0</td>\n",
       "      <td>0.455</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>959</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>531135.5</td>\n",
       "      <td>531135.5</td>\n",
       "      <td>149840746.0</td>\n",
       "      <td>149855154.0</td>\n",
       "      <td>0.98</td>\n",
       "      <td>inversion_paired</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>48.29</td>\n",
       "      <td>2.72</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>14408.0</td>\n",
       "      <td>0.499</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2795</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>30390774.8</td>\n",
       "      <td>30389965.3</td>\n",
       "      <td>211826848.0</td>\n",
       "      <td>211840496.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>inversion_paired</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>8376.11</td>\n",
       "      <td>4.80</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>13648.0</td>\n",
       "      <td>0.556</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7183</td>\n",
       "      <td>1860</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>249179.9</td>\n",
       "      <td>249179.9</td>\n",
       "      <td>5986390.0</td>\n",
       "      <td>6738295.0</td>\n",
       "      <td>0.86</td>\n",
       "      <td>inversion</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2710</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>61.41</td>\n",
       "      <td>33.66</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>751905.0</td>\n",
       "      <td>0.189</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7185</td>\n",
       "      <td>2402</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>218808.3</td>\n",
       "      <td>218808.3</td>\n",
       "      <td>5986390.0</td>\n",
       "      <td>6738295.0</td>\n",
       "      <td>0.86</td>\n",
       "      <td>inversion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2710</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>31.70</td>\n",
       "      <td>49.95</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>751905.0</td>\n",
       "      <td>0.192</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  SmapID QryContigID chr RefcontigID2 QryStartPos   QryEndPos  RefStartPos  \\\n",
       "0    707          42   1            1  16633490.1  16633490.1  103711691.0   \n",
       "1    959         211   1            1    531135.5    531135.5  149840746.0   \n",
       "2   2795          12   2            2  30390774.8  30389965.3  211826848.0   \n",
       "3   7183        1860   7            7    249179.9    249179.9    5986390.0   \n",
       "4   7185        2402   7            7    218808.3    218808.3    5986390.0   \n",
       "\n",
       "     RefEndPos Confidence              Type  ... Genotype GenotypeGroup  \\\n",
       "0  103734414.0       0.99  inversion_paired  ...        1            -1   \n",
       "1  149855154.0       0.98  inversion_paired  ...        1            -1   \n",
       "2  211840496.0        0.9  inversion_paired  ...        1            -1   \n",
       "3    6738295.0       0.86         inversion  ...        1          2710   \n",
       "4    6738295.0       0.86         inversion  ...        2          2710   \n",
       "\n",
       "  RawConfidence RawConfidenceLeft RawConfidenceRight RawConfidenceCenter  \\\n",
       "0         -1.00          16069.91               5.25               -1.00   \n",
       "1         -1.00             48.29               2.72               -1.00   \n",
       "2         -1.00           8376.11               4.80               -1.00   \n",
       "3         -1.00             61.41              33.66               -1.00   \n",
       "4         -1.00             31.70              49.95               -1.00   \n",
       "\n",
       "     SVsize SVfreq orientation Sample_ID  \n",
       "0   22723.0  0.455         NA    GM24385  \n",
       "1   14408.0  0.499         NA    GM24385  \n",
       "2   13648.0  0.556         NA    GM24385  \n",
       "3  751905.0  0.189         NA    GM24385  \n",
       "4  751905.0  0.192         NA    GM24385  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load compiled sv list\n",
    "sig_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_filtered_out.csv\", dtype=object,  index_col=False)\n",
    "display(sig_df.head())\n",
    "sig_df = sig_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "234 has gaps, and 6233 has no gaps, for a total of 6467 SVs\n"
     ]
    }
   ],
   "source": [
    "# remove indel_ngaps\n",
    "# True if overlaps gaps, false if not, remove Falses\n",
    "sig_df[\"has_gaps\"] = sig_df.apply(lambda x: filter_nogaps(x[\"Type\"], x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "nogaps_df = sig_df.loc[sig_df[\"has_gaps\"] == False]\n",
    "\n",
    "num_true = len(sig_df[sig_df[\"has_gaps\"] == True])\n",
    "num_false = len(sig_df[sig_df[\"has_gaps\"] == False])\n",
    "\n",
    "print(f\"{num_true} has gaps, and {num_false} has no gaps, for a total of {num_true + num_false} SVs\")\n",
    "\n",
    "# out_file(nogaps_df, \"case_cleaned_out\")\n",
    "nogaps_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_nogaps_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "From 6233 cleaned SVs, 2202 SVs do not overlap any gene, while 4031 do.\n"
     ]
    }
   ],
   "source": [
    "# find overlap genes matching hg38knownCanonical.bed\n",
    "nogaps_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_nogaps_out.csv\", dtype=object,  index_col=False)\n",
    "nogaps_df = nogaps_df.astype({'RefStartPos':'float64','RefEndPos':'float64','chr':'int'})\n",
    "nogaps_df.dtypes\n",
    "\n",
    "nogaps_df[\"OverlapGenes\"] = nogaps_df.apply(lambda x: filter_genes(bed_df, x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "\n",
    "no_overlaps = len(nogaps_df.loc[nogaps_df[\"OverlapGenes\"] == \"-\"])\n",
    "overlaps = len(nogaps_df.loc[nogaps_df[\"OverlapGenes\"] != \"-\"])\n",
    "print(f\"From {no_overlaps + overlaps} cleaned SVs, {no_overlaps} SVs do not overlap any gene, while {overlaps} do.\")\n",
    "\n",
    "# out_file(nogaps_df, \"control_cleaned_out\")\n",
    "nogaps_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_cleaned_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_cleaned_out.csv\", dtype=object,  index_col=False)\n",
    "sv_df = sv_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'SVsize':'float64', 'chr':'int'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_df['OverlapGenes'] = sv_df.apply(lambda x: parse_sorted_unique_genes(x['OverlapGenes']), axis=1)\n",
    "sv_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_cleaned_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((6233, 30),\n",
       " insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_df.shape, sv_df['Type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "39\n",
      "60\n",
      "101\n",
      "126\n",
      "142\n",
      "180\n",
      "213\n",
      "226\n",
      "237\n",
      "265\n",
      "285\n",
      "311\n",
      "325\n",
      "362\n",
      "384\n",
      "407\n",
      "417\n",
      "439\n",
      "463\n",
      "479\n",
      "486\n",
      "509\n",
      "544\n",
      "558\n",
      "575\n",
      "576\n",
      "594\n",
      "624\n",
      "625\n",
      "635\n",
      "647\n",
      "661\n",
      "682\n",
      "683\n",
      "701\n",
      "710\n",
      "733\n",
      "750\n",
      "759\n",
      "768\n",
      "30.0%/ 100% done\n",
      "791\n",
      "825\n",
      "832\n",
      "841\n",
      "853\n",
      "878\n",
      "885\n",
      "891\n",
      "908\n",
      "930\n",
      "931\n",
      "933\n",
      "948\n",
      "952\n",
      "968\n",
      "991\n",
      "1003\n",
      "1004\n",
      "1014\n",
      "1021\n",
      "1028\n",
      "1029\n",
      "1038\n",
      "1046\n",
      "1052\n",
      "1054\n",
      "1065\n",
      "1077\n",
      "1078\n",
      "1082\n",
      "1091\n",
      "1106\n",
      "1114\n",
      "1120\n",
      "1128\n",
      "1146\n",
      "1151\n",
      "1153\n",
      "1156\n",
      "1157\n",
      "1169\n",
      "1180\n",
      "1186\n",
      "1188\n",
      "1194\n",
      "1210\n",
      "1215\n",
      "1218\n",
      "1228\n",
      "1239\n",
      "1247\n",
      "1249\n",
      "1250\n",
      "1259\n",
      "1272\n",
      "1277\n",
      "1296\n",
      "1299\n",
      "1366\n",
      "1368\n",
      "1371\n",
      "1373\n",
      "1374\n",
      "1375\n",
      "1376\n",
      "1377\n",
      "1378\n",
      "1379\n",
      "1380\n",
      "1381\n",
      "1383\n",
      "1384\n",
      "1385\n",
      "1386\n",
      "1387\n",
      "1388\n",
      "1389\n",
      "1390\n",
      "1391\n",
      "1392\n",
      "1393\n",
      "1394\n",
      "1395\n",
      "1396\n",
      "1397\n",
      "1398\n",
      "1399\n",
      "1400\n",
      "1401\n",
      "1402\n",
      "1404\n",
      "1405\n",
      "1406\n",
      "1407\n",
      "1408\n",
      "1409\n",
      "1410\n",
      "1411\n",
      "1412\n",
      "1413\n",
      "1414\n",
      "1415\n",
      "1416\n",
      "1417\n",
      "1418\n",
      "1419\n",
      "1420\n",
      "1421\n",
      "1422\n",
      "1423\n",
      "1424\n",
      "1425\n",
      "1426\n",
      "1427\n",
      "1428\n",
      "1429\n",
      "1430\n",
      "1431\n",
      "1432\n",
      "1433\n",
      "1434\n",
      "1435\n",
      "1436\n",
      "1437\n",
      "1438\n",
      "1439\n",
      "1440\n",
      "1441\n",
      "1442\n",
      "1443\n",
      "1444\n",
      "1445\n",
      "1446\n",
      "1447\n",
      "1448\n",
      "1449\n",
      "1450\n",
      "1451\n",
      "1452\n",
      "1453\n",
      "1454\n",
      "1455\n",
      "1456\n",
      "1457\n",
      "1458\n",
      "1459\n",
      "1460\n",
      "1461\n",
      "1462\n",
      "1463\n",
      "1464\n",
      "1465\n",
      "1466\n",
      "1467\n",
      "1468\n",
      "1469\n",
      "1470\n",
      "1471\n",
      "1472\n",
      "1473\n",
      "1474\n",
      "1475\n",
      "1476\n",
      "1477\n",
      "1478\n",
      "1479\n",
      "1480\n",
      "1481\n",
      "1482\n",
      "1483\n",
      "1484\n",
      "1485\n",
      "1486\n",
      "1487\n",
      "1488\n",
      "1489\n",
      "1490\n",
      "1491\n",
      "1492\n",
      "1493\n",
      "1494\n",
      "1495\n",
      "1496\n",
      "1497\n",
      "1498\n",
      "1499\n",
      "1500\n",
      "1501\n",
      "1502\n",
      "1504\n",
      "1505\n",
      "1506\n",
      "1507\n",
      "1508\n",
      "1509\n",
      "1510\n",
      "1511\n",
      "1512\n",
      "1513\n",
      "1514\n",
      "1515\n",
      "1516\n",
      "1517\n",
      "1518\n",
      "1519\n",
      "1520\n",
      "1521\n",
      "1522\n",
      "1523\n",
      "1524\n",
      "1525\n",
      "1526\n",
      "1527\n",
      "1528\n",
      "1529\n",
      "1530\n",
      "1531\n",
      "1532\n",
      "1533\n",
      "1534\n",
      "1535\n",
      "1536\n",
      "60.0%/ 100% done\n",
      "1537\n",
      "1538\n",
      "1539\n",
      "1540\n",
      "1541\n",
      "1542\n",
      "1543\n",
      "1544\n",
      "1545\n",
      "1546\n",
      "1547\n",
      "1548\n",
      "1549\n",
      "1550\n",
      "1551\n",
      "1553\n",
      "1554\n",
      "1555\n",
      "1556\n",
      "1557\n",
      "1558\n",
      "1559\n",
      "1560\n",
      "1561\n",
      "1563\n",
      "1565\n",
      "1567\n",
      "1568\n",
      "1569\n",
      "1570\n",
      "1571\n",
      "1572\n",
      "1573\n",
      "1574\n",
      "1575\n",
      "1576\n",
      "1577\n",
      "1578\n",
      "1579\n",
      "1580\n",
      "1581\n",
      "1582\n",
      "1583\n",
      "1584\n",
      "1585\n",
      "1586\n",
      "1587\n",
      "1588\n",
      "1589\n",
      "1590\n",
      "1591\n",
      "1592\n",
      "1593\n",
      "1594\n",
      "1595\n",
      "1596\n",
      "1597\n",
      "1598\n",
      "1599\n",
      "1600\n",
      "1601\n",
      "1602\n",
      "1604\n",
      "1605\n",
      "1606\n",
      "1607\n",
      "1608\n",
      "1609\n",
      "1610\n",
      "1611\n",
      "1612\n",
      "1613\n",
      "1614\n",
      "1615\n",
      "1616\n",
      "1617\n",
      "1618\n",
      "1619\n",
      "1620\n",
      "1621\n",
      "1622\n",
      "1623\n",
      "1624\n",
      "1625\n",
      "1626\n",
      "1627\n",
      "1628\n",
      "1629\n",
      "1630\n",
      "1631\n",
      "1632\n",
      "1633\n",
      "1634\n",
      "1635\n",
      "1636\n",
      "1637\n",
      "1638\n",
      "1639\n",
      "1640\n",
      "1641\n",
      "1642\n",
      "1643\n",
      "1644\n",
      "1645\n",
      "1646\n",
      "1647\n",
      "1648\n",
      "1649\n",
      "1650\n",
      "1651\n",
      "1652\n",
      "1653\n",
      "1654\n",
      "1655\n",
      "1656\n",
      "1657\n",
      "1658\n",
      "1659\n",
      "1660\n",
      "1661\n",
      "1662\n",
      "1663\n",
      "1664\n",
      "1665\n",
      "1666\n",
      "1667\n",
      "1668\n",
      "1669\n",
      "1670\n",
      "1671\n",
      "1672\n",
      "1673\n",
      "1674\n",
      "1675\n",
      "1676\n",
      "1677\n",
      "1678\n",
      "1679\n",
      "1680\n",
      "1681\n",
      "1682\n",
      "1683\n",
      "1684\n",
      "1685\n",
      "1686\n",
      "1687\n",
      "1688\n",
      "1689\n",
      "1690\n",
      "1691\n",
      "1693\n",
      "1694\n",
      "1695\n",
      "1696\n",
      "1697\n",
      "1698\n",
      "1699\n",
      "1701\n",
      "1702\n",
      "1703\n",
      "1704\n",
      "1705\n",
      "1706\n",
      "1707\n",
      "1708\n",
      "1709\n",
      "1710\n",
      "1711\n",
      "1712\n",
      "1713\n",
      "1714\n",
      "1715\n",
      "1716\n",
      "1717\n",
      "1718\n",
      "1719\n",
      "1720\n",
      "1721\n",
      "1722\n",
      "1723\n",
      "1724\n",
      "1725\n",
      "1726\n",
      "1727\n",
      "1728\n",
      "1729\n",
      "1730\n",
      "1732\n",
      "1733\n",
      "1734\n",
      "1735\n",
      "1736\n",
      "1737\n",
      "1738\n",
      "1739\n",
      "1740\n",
      "1741\n",
      "1742\n",
      "1743\n",
      "1744\n",
      "1745\n",
      "1746\n",
      "1747\n",
      "1748\n",
      "1750\n",
      "1751\n",
      "1752\n",
      "1753\n",
      "1754\n",
      "1755\n",
      "1756\n",
      "1757\n",
      "1758\n",
      "1759\n",
      "1760\n",
      "1761\n",
      "1762\n",
      "1763\n",
      "1764\n",
      "1765\n",
      "1766\n",
      "1767\n",
      "1768\n",
      "1769\n",
      "1770\n",
      "1771\n",
      "1772\n",
      "1773\n",
      "1774\n",
      "1775\n",
      "1776\n",
      "1777\n",
      "1778\n",
      "1779\n",
      "1780\n",
      "1781\n",
      "1782\n",
      "1783\n",
      "1784\n",
      "1785\n",
      "1786\n",
      "1787\n",
      "1788\n",
      "1789\n",
      "1790\n",
      "1791\n",
      "1792\n",
      "70.0%/ 100% done\n",
      "1793\n",
      "1794\n",
      "1795\n",
      "1796\n",
      "1797\n",
      "1798\n",
      "1799\n",
      "1800\n",
      "1801\n",
      "1802\n",
      "1803\n",
      "1804\n",
      "1805\n",
      "1806\n",
      "1807\n",
      "1808\n",
      "1809\n",
      "1810\n",
      "1811\n",
      "1812\n",
      "1813\n",
      "1814\n",
      "1815\n",
      "1816\n",
      "1817\n",
      "1818\n",
      "1819\n",
      "1820\n",
      "1821\n",
      "1822\n",
      "1824\n",
      "1825\n",
      "1826\n",
      "1827\n",
      "1828\n",
      "1829\n",
      "1830\n",
      "1831\n",
      "1832\n",
      "1833\n",
      "1835\n",
      "1836\n",
      "1837\n",
      "1839\n",
      "1840\n",
      "1841\n",
      "1842\n",
      "1843\n",
      "1844\n",
      "1845\n",
      "1846\n",
      "1847\n",
      "1848\n",
      "1849\n",
      "1850\n",
      "1851\n",
      "1852\n",
      "1853\n",
      "1854\n",
      "1855\n",
      "1856\n",
      "1857\n",
      "1858\n",
      "1859\n",
      "1860\n",
      "1861\n",
      "1862\n",
      "1863\n",
      "1864\n",
      "1865\n",
      "1866\n",
      "1867\n",
      "1868\n",
      "1869\n",
      "1870\n",
      "1871\n",
      "1872\n",
      "1873\n",
      "1875\n",
      "1876\n",
      "1877\n",
      "1878\n",
      "1879\n",
      "1880\n",
      "1881\n",
      "1882\n",
      "1883\n",
      "1884\n",
      "1885\n",
      "1887\n",
      "1889\n",
      "1890\n",
      "1891\n",
      "1892\n",
      "1893\n",
      "1894\n",
      "1895\n",
      "1896\n",
      "1897\n",
      "1898\n",
      "1899\n",
      "1900\n",
      "1901\n",
      "1902\n",
      "1903\n",
      "1904\n",
      "1905\n",
      "1906\n",
      "1907\n",
      "1908\n",
      "1909\n",
      "1910\n",
      "1911\n",
      "1912\n",
      "1913\n",
      "1915\n",
      "1916\n",
      "1917\n",
      "1918\n",
      "1919\n",
      "1920\n",
      "1921\n",
      "1922\n",
      "1923\n",
      "1924\n",
      "1925\n",
      "1926\n",
      "1927\n",
      "1928\n",
      "1929\n",
      "1930\n",
      "1931\n",
      "1932\n",
      "1933\n",
      "1934\n",
      "1935\n",
      "1936\n",
      "1937\n",
      "1938\n",
      "1939\n",
      "1940\n",
      "1941\n",
      "1942\n",
      "1943\n",
      "1944\n",
      "1945\n",
      "1946\n",
      "1947\n",
      "1949\n",
      "1950\n",
      "1951\n",
      "1952\n",
      "1953\n",
      "1954\n",
      "1955\n",
      "1956\n",
      "1957\n",
      "1958\n",
      "1959\n",
      "1960\n",
      "1961\n",
      "1962\n",
      "1963\n",
      "1964\n",
      "1965\n",
      "1966\n",
      "1967\n",
      "1968\n",
      "1969\n",
      "1970\n",
      "1971\n",
      "1972\n",
      "1973\n",
      "1974\n",
      "1975\n",
      "1976\n",
      "1977\n",
      "1978\n",
      "1979\n",
      "1980\n",
      "1981\n",
      "1982\n",
      "1983\n",
      "1984\n",
      "1985\n",
      "1986\n",
      "1987\n",
      "1988\n",
      "1989\n",
      "1990\n",
      "1991\n",
      "1992\n",
      "1993\n",
      "1994\n",
      "1995\n",
      "1997\n",
      "1998\n",
      "1999\n",
      "2000\n",
      "2001\n",
      "2002\n",
      "2003\n",
      "2004\n",
      "2005\n",
      "2006\n",
      "2007\n",
      "2008\n",
      "2009\n",
      "2010\n",
      "2011\n",
      "2013\n",
      "2014\n",
      "2015\n",
      "2016\n",
      "2017\n",
      "2018\n",
      "2019\n",
      "2020\n",
      "2022\n",
      "2023\n",
      "2024\n",
      "2025\n",
      "2026\n",
      "2028\n",
      "2029\n",
      "2030\n",
      "2031\n",
      "2032\n",
      "2033\n",
      "2034\n",
      "2035\n",
      "2036\n",
      "2037\n",
      "2038\n",
      "2039\n",
      "2040\n",
      "2041\n",
      "2042\n",
      "2043\n",
      "2044\n",
      "2046\n",
      "2047\n",
      "2048\n",
      "80.0%/ 100% done\n",
      "2049\n",
      "2050\n",
      "2051\n",
      "2053\n",
      "2054\n",
      "2055\n",
      "2056\n",
      "2057\n",
      "2058\n",
      "2059\n",
      "2060\n",
      "2061\n",
      "2062\n",
      "2063\n",
      "2064\n",
      "2065\n",
      "2066\n",
      "2067\n",
      "2068\n",
      "2069\n",
      "2070\n",
      "2071\n",
      "2072\n",
      "2073\n",
      "2074\n",
      "2075\n",
      "2076\n",
      "2077\n",
      "2078\n",
      "2079\n",
      "2080\n",
      "2081\n",
      "2083\n",
      "2084\n",
      "2085\n",
      "2086\n",
      "2087\n",
      "2088\n",
      "2089\n",
      "2090\n",
      "2091\n",
      "2092\n",
      "2093\n",
      "2094\n",
      "2095\n",
      "2096\n",
      "2097\n",
      "2098\n",
      "2099\n",
      "2100\n",
      "2101\n",
      "2102\n",
      "2103\n",
      "2104\n",
      "2105\n",
      "2106\n",
      "2107\n",
      "2108\n",
      "2109\n",
      "2110\n",
      "2111\n",
      "2112\n",
      "2113\n",
      "2114\n",
      "2115\n",
      "2116\n",
      "2117\n",
      "2118\n",
      "2119\n",
      "2120\n",
      "2121\n",
      "2122\n",
      "2123\n",
      "2125\n",
      "2126\n",
      "2127\n",
      "2128\n",
      "2129\n",
      "2130\n",
      "2131\n",
      "2132\n",
      "2133\n",
      "2134\n",
      "2135\n",
      "2136\n",
      "2137\n",
      "2138\n",
      "2139\n",
      "2140\n",
      "2141\n",
      "2142\n",
      "2143\n",
      "2144\n",
      "2145\n",
      "2146\n",
      "2147\n",
      "2148\n",
      "2149\n",
      "2150\n",
      "2151\n",
      "2152\n",
      "2153\n",
      "2154\n",
      "2155\n",
      "2156\n",
      "2157\n",
      "2158\n",
      "2159\n",
      "2160\n",
      "2161\n",
      "2162\n",
      "2163\n",
      "2164\n",
      "2165\n",
      "2166\n",
      "2167\n",
      "2168\n",
      "2169\n",
      "2170\n",
      "2171\n",
      "2172\n",
      "2173\n",
      "2174\n",
      "2175\n",
      "2177\n",
      "2178\n",
      "2179\n",
      "2180\n",
      "2181\n",
      "2182\n",
      "2183\n",
      "2184\n",
      "2185\n",
      "2186\n",
      "2187\n",
      "2189\n",
      "2190\n",
      "2191\n",
      "2192\n",
      "2193\n",
      "2194\n",
      "2195\n",
      "2196\n",
      "2197\n",
      "2198\n",
      "2199\n",
      "2200\n",
      "2201\n",
      "2202\n",
      "2203\n",
      "2204\n",
      "2205\n",
      "2206\n",
      "2207\n",
      "2208\n",
      "2209\n",
      "2211\n",
      "2212\n",
      "2213\n",
      "2214\n",
      "2215\n",
      "2216\n",
      "2217\n",
      "2218\n",
      "2219\n",
      "2220\n",
      "2221\n",
      "2222\n",
      "2223\n",
      "2224\n",
      "2225\n",
      "2226\n",
      "2227\n",
      "2228\n",
      "2229\n",
      "2230\n",
      "2231\n",
      "2232\n",
      "2233\n",
      "2234\n",
      "2235\n",
      "2236\n",
      "2237\n",
      "2238\n",
      "2239\n",
      "2240\n",
      "2241\n",
      "2242\n",
      "2243\n",
      "2244\n",
      "2245\n",
      "2246\n",
      "2247\n",
      "2248\n",
      "2249\n",
      "2250\n",
      "2251\n",
      "2252\n",
      "2253\n",
      "2254\n",
      "2255\n",
      "2256\n",
      "2257\n",
      "2258\n",
      "2259\n",
      "2260\n",
      "2262\n",
      "2263\n",
      "2264\n",
      "2265\n",
      "2266\n",
      "2267\n",
      "2268\n",
      "2269\n",
      "2270\n",
      "2271\n",
      "2272\n",
      "2273\n",
      "2274\n",
      "2275\n",
      "2276\n",
      "2277\n",
      "2278\n",
      "2279\n",
      "2280\n",
      "2281\n",
      "2282\n",
      "2283\n",
      "2284\n",
      "2285\n",
      "2286\n",
      "2287\n",
      "2288\n",
      "2289\n",
      "2290\n",
      "2291\n",
      "2292\n",
      "2293\n",
      "2295\n",
      "2296\n",
      "2297\n",
      "2298\n",
      "2299\n",
      "2300\n",
      "2301\n",
      "2302\n",
      "2303\n",
      "2304\n",
      "90.0%/ 100% done\n",
      "2305\n",
      "2306\n",
      "2307\n",
      "2308\n",
      "2309\n",
      "2310\n",
      "2311\n",
      "2312\n",
      "2313\n",
      "2314\n",
      "2315\n",
      "2316\n",
      "2317\n",
      "2318\n",
      "2319\n",
      "2320\n",
      "2321\n",
      "2322\n",
      "2323\n",
      "2324\n",
      "2325\n",
      "2326\n",
      "2327\n",
      "2328\n",
      "2329\n",
      "2330\n",
      "2331\n",
      "2332\n",
      "2333\n",
      "2334\n",
      "2335\n",
      "2336\n",
      "2337\n",
      "2338\n",
      "2339\n",
      "2340\n",
      "2341\n",
      "2342\n",
      "2343\n",
      "2344\n",
      "2345\n",
      "2346\n",
      "2347\n",
      "2348\n",
      "2349\n",
      "2350\n",
      "2351\n",
      "2352\n",
      "2353\n",
      "2354\n",
      "2355\n",
      "2356\n",
      "2357\n",
      "2358\n",
      "2359\n",
      "2360\n",
      "2361\n",
      "2362\n",
      "2363\n",
      "2364\n",
      "2365\n",
      "2366\n",
      "2367\n",
      "2368\n",
      "2369\n",
      "2370\n",
      "2371\n",
      "2372\n",
      "2373\n",
      "2374\n",
      "2375\n",
      "2376\n",
      "2377\n",
      "2378\n",
      "2379\n",
      "2381\n",
      "2382\n",
      "2383\n",
      "2384\n",
      "2385\n",
      "2386\n",
      "2387\n",
      "2388\n",
      "2389\n",
      "2391\n",
      "2392\n",
      "2393\n",
      "2394\n",
      "2395\n",
      "2396\n",
      "2397\n",
      "2398\n",
      "2399\n",
      "2400\n",
      "2401\n",
      "2402\n",
      "2403\n",
      "2405\n",
      "2406\n",
      "2407\n",
      "2408\n",
      "2409\n",
      "2410\n",
      "2411\n",
      "2412\n",
      "2413\n",
      "2414\n",
      "2415\n",
      "2416\n",
      "2417\n",
      "2418\n",
      "2419\n",
      "2420\n",
      "2421\n",
      "2422\n",
      "2423\n",
      "2424\n",
      "2425\n",
      "2426\n",
      "2427\n",
      "2428\n",
      "2429\n",
      "2430\n",
      "2431\n",
      "2433\n",
      "2434\n",
      "2435\n",
      "2436\n",
      "2438\n",
      "2439\n",
      "2440\n",
      "2441\n",
      "2442\n",
      "2443\n",
      "2444\n",
      "2446\n",
      "2447\n",
      "2448\n",
      "2449\n",
      "2450\n",
      "2451\n",
      "2452\n",
      "2453\n",
      "2454\n",
      "2455\n",
      "2456\n",
      "2457\n",
      "2458\n",
      "2459\n",
      "2460\n",
      "2461\n",
      "2462\n",
      "2463\n",
      "2464\n",
      "2465\n",
      "2466\n",
      "2467\n",
      "2468\n",
      "2469\n",
      "2470\n",
      "2471\n",
      "2472\n",
      "2473\n",
      "2474\n",
      "2475\n",
      "2476\n",
      "2477\n",
      "2478\n",
      "2479\n",
      "2480\n",
      "2481\n",
      "2482\n",
      "2483\n",
      "2484\n",
      "2485\n",
      "2486\n",
      "2487\n",
      "2488\n",
      "2489\n",
      "2490\n",
      "2491\n",
      "2492\n",
      "2493\n",
      "2494\n",
      "2495\n",
      "2496\n",
      "2498\n",
      "2499\n",
      "2500\n",
      "2501\n",
      "2502\n",
      "2503\n",
      "2504\n",
      "2505\n",
      "2506\n",
      "2507\n",
      "2508\n",
      "2509\n",
      "2510\n",
      "2511\n",
      "2512\n",
      "2514\n",
      "2515\n",
      "2516\n",
      "2517\n",
      "2518\n",
      "2520\n",
      "2522\n",
      "2523\n",
      "2524\n",
      "2525\n",
      "2526\n",
      "2528\n",
      "2529\n",
      "2530\n",
      "2531\n",
      "2532\n",
      "2533\n",
      "2534\n",
      "2535\n",
      "2536\n",
      "2537\n",
      "2538\n",
      "2539\n",
      "2540\n",
      "2541\n",
      "2542\n",
      "2543\n",
      "2544\n",
      "2545\n",
      "2546\n",
      "2547\n",
      "2548\n",
      "2549\n",
      "2550\n",
      "2551\n",
      "2552\n",
      "2553\n",
      "2554\n",
      "2555\n",
      "2556\n",
      "2557\n",
      "2558\n",
      "2559\n",
      "2560\n",
      "100.0%/ 100% done\n",
      "2562\n",
      "2563\n",
      "2564\n",
      "2565\n",
      "2566\n",
      "2567\n",
      "2568\n",
      "2569\n",
      "2570\n",
      "2571\n",
      "2572\n",
      "2573\n",
      "2574\n",
      "2575\n",
      "2576\n",
      "2577\n",
      "2578\n",
      "2579\n",
      "2580\n",
      "2581\n",
      "2582\n",
      "2583\n",
      "2584\n",
      "2585\n",
      "2586\n",
      "2587\n",
      "2588\n",
      "2589\n",
      "2590\n",
      "2591\n",
      "2592\n",
      "2593\n",
      "2594\n",
      "2595\n",
      "2596\n",
      "2597\n",
      "2598\n",
      "2599\n",
      "2600\n",
      "2602\n",
      "2603\n",
      "2604\n",
      "2606\n",
      "2607\n",
      "2608\n",
      "2609\n",
      "2610\n",
      "2611\n",
      "2612\n",
      "2614\n",
      "2615\n",
      "2616\n",
      "2617\n",
      "2618\n",
      "2619\n",
      "2620\n",
      "2621\n",
      "2622\n",
      "2624\n",
      "2625\n",
      "2626\n",
      "2627\n",
      "2628\n",
      "2629\n",
      "2630\n",
      "2631\n",
      "2632\n",
      "2633\n",
      "2634\n",
      "2635\n",
      "2636\n",
      "2638\n",
      "2639\n",
      "2640\n",
      "2641\n",
      "2642\n",
      "2643\n",
      "2644\n",
      "2645\n",
      "2646\n",
      "2647\n",
      "2648\n",
      "2649\n",
      "2650\n",
      "2651\n",
      "2652\n",
      "2653\n",
      "2654\n",
      "2655\n",
      "2656\n",
      "2657\n",
      "2658\n",
      "2659\n",
      "2660\n",
      "2661\n",
      "2662\n",
      "2663\n",
      "2664\n",
      "2665\n",
      "2666\n",
      "2668\n",
      "2670\n",
      "2671\n",
      "2672\n",
      "2673\n",
      "2674\n",
      "2675\n",
      "2676\n",
      "2677\n",
      "2678\n",
      "2679\n",
      "2680\n",
      "2681\n",
      "2682\n",
      "2683\n",
      "2684\n",
      "2685\n",
      "2686\n",
      "2687\n",
      "2688\n",
      "2689\n",
      "2690\n",
      "2691\n",
      "2692\n",
      "2693\n",
      "2694\n",
      "2695\n",
      "2697\n",
      "2698\n",
      "2699\n",
      "2700\n",
      "2701\n",
      "2702\n",
      "2703\n",
      "2704\n",
      "2705\n",
      "2706\n",
      "2707\n",
      "2708\n",
      "2709\n",
      "2711\n",
      "2713\n",
      "2714\n",
      "2715\n",
      "2716\n",
      "2717\n",
      "2718\n",
      "2719\n",
      "2720\n",
      "2721\n",
      "2722\n",
      "2723\n",
      "2724\n",
      "2725\n",
      "2726\n",
      "2727\n",
      "2730\n",
      "2731\n",
      "2732\n",
      "2733\n",
      "2734\n",
      "2735\n",
      "2736\n",
      "2737\n",
      "2738\n",
      "2742\n",
      "2747\n",
      "2748\n",
      "2749\n",
      "2750\n",
      "2751\n",
      "2752\n",
      "2753\n",
      "2754\n",
      "2755\n",
      "2758\n",
      "2759\n",
      "2760\n",
      "2763\n",
      "2764\n",
      "2765\n",
      "2767\n",
      "2769\n",
      "2770\n",
      "2771\n",
      "2772\n",
      "2773\n",
      "2774\n",
      "2775\n",
      "2776\n",
      "2777\n",
      "2778\n",
      "2779\n",
      "2780\n",
      "2781\n",
      "2782\n",
      "2783\n",
      "2784\n",
      "2785\n",
      "2786\n",
      "2787\n",
      "2788\n",
      "2789\n",
      "2790\n",
      "2791\n",
      "2792\n",
      "2793\n",
      "2794\n",
      "2795\n",
      "2796\n",
      "2797\n",
      "2798\n",
      "2800\n",
      "2801\n",
      "2802\n",
      "2803\n",
      "2804\n",
      "2805\n",
      "2806\n",
      "2807\n",
      "2808\n",
      "2809\n",
      "2810\n",
      "2811\n",
      "2812\n",
      "2813\n",
      "2814\n",
      "2815\n",
      "2816\n",
      "110.00000000000001%/ 100% done\n",
      "2817\n",
      "2818\n",
      "2820\n",
      "2821\n",
      "2822\n",
      "2823\n",
      "2824\n",
      "2825\n",
      "2826\n",
      "2827\n",
      "2828\n",
      "2829\n",
      "2830\n",
      "2831\n",
      "2832\n",
      "2833\n",
      "2834\n",
      "2835\n",
      "2836\n",
      "2837\n",
      "2838\n",
      "2839\n",
      "2840\n",
      "2841\n",
      "2842\n",
      "2843\n",
      "2844\n",
      "2846\n",
      "2847\n",
      "2848\n",
      "2849\n",
      "2850\n",
      "2851\n",
      "2852\n",
      "2853\n",
      "2854\n",
      "2855\n",
      "2856\n",
      "2857\n",
      "2858\n",
      "2859\n",
      "2860\n",
      "2861\n",
      "2862\n",
      "2864\n",
      "2865\n",
      "2866\n",
      "2867\n",
      "2868\n",
      "2869\n",
      "2870\n",
      "2871\n",
      "2872\n",
      "2873\n",
      "2874\n",
      "2875\n",
      "2876\n",
      "2877\n",
      "2878\n",
      "2879\n",
      "2880\n",
      "2881\n",
      "2882\n",
      "2884\n",
      "2885\n",
      "2886\n",
      "2887\n",
      "2888\n",
      "2890\n",
      "2891\n",
      "2892\n",
      "2893\n",
      "2894\n",
      "2895\n",
      "2896\n",
      "2899\n",
      "2900\n",
      "2901\n",
      "2902\n",
      "2903\n",
      "2904\n",
      "2905\n",
      "2906\n",
      "2907\n",
      "2908\n",
      "2909\n",
      "2910\n",
      "2911\n",
      "2912\n",
      "2913\n",
      "2914\n",
      "2915\n",
      "2916\n",
      "2917\n",
      "2919\n",
      "2920\n",
      "2921\n",
      "2922\n",
      "2923\n",
      "2924\n",
      "2925\n",
      "2926\n",
      "2927\n",
      "2928\n",
      "2929\n",
      "2930\n",
      "2931\n",
      "2932\n",
      "2933\n",
      "2934\n",
      "2935\n",
      "2936\n",
      "2937\n",
      "2938\n",
      "2939\n",
      "2940\n",
      "2942\n",
      "2943\n",
      "2944\n",
      "2945\n",
      "2946\n",
      "2947\n",
      "2949\n",
      "2950\n",
      "2951\n",
      "2952\n",
      "2953\n",
      "2954\n",
      "2955\n",
      "2956\n",
      "2957\n",
      "2958\n",
      "2959\n",
      "2960\n",
      "2961\n",
      "2962\n",
      "2963\n",
      "2964\n",
      "2965\n",
      "2966\n",
      "2967\n",
      "2968\n",
      "2969\n",
      "2970\n",
      "2971\n",
      "2972\n",
      "2973\n",
      "2974\n",
      "2975\n",
      "2976\n",
      "2977\n",
      "2978\n",
      "2979\n",
      "2980\n",
      "2981\n",
      "2982\n",
      "2983\n",
      "2984\n",
      "2985\n",
      "2986\n",
      "2987\n",
      "2989\n",
      "2990\n",
      "2991\n",
      "2992\n",
      "2993\n",
      "2994\n",
      "2995\n",
      "2996\n",
      "2997\n",
      "2998\n",
      "2999\n",
      "3000\n",
      "3002\n",
      "3003\n",
      "3004\n",
      "3005\n",
      "3006\n",
      "3007\n",
      "3008\n",
      "3009\n",
      "3010\n",
      "3011\n",
      "3012\n",
      "3013\n",
      "3014\n",
      "3015\n",
      "3016\n",
      "3017\n",
      "3018\n",
      "3019\n",
      "3020\n",
      "3021\n",
      "3022\n",
      "3023\n",
      "3024\n",
      "3025\n",
      "3026\n",
      "3027\n",
      "3028\n",
      "3029\n",
      "3030\n",
      "3031\n",
      "3032\n",
      "3033\n",
      "3034\n",
      "3035\n",
      "3036\n",
      "3037\n",
      "3038\n",
      "3039\n",
      "3040\n",
      "3041\n",
      "3042\n",
      "3043\n",
      "3044\n",
      "3045\n",
      "3046\n",
      "3047\n",
      "3048\n",
      "3049\n",
      "3050\n",
      "3051\n",
      "3052\n",
      "3053\n",
      "3054\n",
      "3055\n",
      "3056\n",
      "3058\n",
      "3059\n",
      "3060\n",
      "3061\n",
      "3062\n",
      "3063\n",
      "3065\n",
      "3067\n",
      "3068\n",
      "3069\n",
      "3070\n",
      "3072\n",
      "120.0%/ 100% done\n",
      "3073\n",
      "3074\n",
      "3075\n",
      "3076\n",
      "3077\n",
      "3078\n",
      "3079\n",
      "3080\n",
      "3081\n",
      "3082\n",
      "3083\n",
      "3084\n",
      "3085\n",
      "3086\n",
      "3087\n",
      "3088\n",
      "3089\n",
      "3090\n",
      "3091\n",
      "3092\n",
      "3093\n",
      "3094\n",
      "3095\n",
      "3096\n",
      "3097\n",
      "3098\n",
      "3099\n",
      "3100\n",
      "3101\n",
      "3102\n",
      "3103\n",
      "3104\n",
      "3105\n",
      "3106\n",
      "3107\n",
      "3108\n",
      "3109\n",
      "3111\n",
      "3112\n",
      "3113\n",
      "3114\n",
      "3115\n",
      "3117\n",
      "3118\n",
      "3119\n",
      "3120\n",
      "3121\n",
      "3122\n",
      "3123\n",
      "3124\n",
      "3126\n",
      "3127\n",
      "3128\n",
      "3129\n",
      "3131\n",
      "3132\n",
      "3133\n",
      "3134\n",
      "3135\n",
      "3136\n",
      "3137\n",
      "3138\n",
      "3139\n",
      "3140\n",
      "3141\n",
      "3142\n",
      "3143\n",
      "3144\n",
      "3145\n",
      "3146\n",
      "3147\n",
      "3148\n",
      "3149\n",
      "3150\n",
      "3151\n",
      "3153\n",
      "3154\n",
      "3155\n",
      "3156\n",
      "3157\n",
      "3158\n",
      "3159\n",
      "3160\n",
      "3161\n",
      "3163\n",
      "3164\n",
      "3165\n",
      "3167\n",
      "3169\n",
      "3170\n",
      "3171\n",
      "3172\n",
      "3173\n",
      "3174\n",
      "3175\n",
      "3176\n",
      "3177\n",
      "3178\n",
      "3179\n",
      "3180\n",
      "3181\n",
      "3182\n",
      "3183\n",
      "3184\n",
      "3185\n",
      "3186\n",
      "3187\n",
      "3188\n",
      "3189\n",
      "3190\n",
      "3191\n",
      "3192\n",
      "3193\n",
      "3194\n",
      "3195\n",
      "3196\n",
      "3197\n",
      "3198\n",
      "3199\n",
      "3200\n",
      "3201\n",
      "3202\n",
      "3203\n",
      "3204\n",
      "3205\n",
      "3206\n",
      "3207\n",
      "3209\n",
      "3210\n",
      "3211\n",
      "3212\n",
      "3213\n",
      "3214\n",
      "3215\n",
      "3216\n",
      "3217\n",
      "3218\n",
      "3219\n",
      "3220\n",
      "3221\n",
      "3222\n",
      "3223\n",
      "3224\n",
      "3225\n",
      "3226\n",
      "3227\n",
      "3228\n",
      "3229\n",
      "3230\n",
      "3231\n",
      "3232\n",
      "3233\n",
      "3234\n",
      "3235\n",
      "3236\n",
      "3237\n",
      "3238\n",
      "3239\n",
      "3240\n",
      "3241\n",
      "3243\n",
      "3244\n",
      "3245\n",
      "3246\n",
      "3247\n",
      "3248\n",
      "3249\n",
      "3250\n",
      "3251\n",
      "3252\n",
      "3253\n",
      "3254\n",
      "3255\n",
      "3256\n",
      "3257\n",
      "3258\n",
      "3259\n",
      "3260\n",
      "3261\n",
      "3262\n",
      "3263\n",
      "3264\n",
      "3265\n",
      "3266\n",
      "3267\n",
      "3268\n",
      "3269\n",
      "3270\n",
      "3271\n",
      "3272\n",
      "3273\n",
      "3274\n",
      "3275\n",
      "3276\n",
      "3277\n",
      "3278\n",
      "3279\n",
      "3280\n",
      "3281\n",
      "3282\n",
      "3283\n",
      "3284\n",
      "3285\n",
      "3286\n",
      "3287\n",
      "3288\n",
      "3289\n",
      "3290\n",
      "3291\n",
      "3293\n",
      "3294\n",
      "3295\n",
      "3296\n",
      "3297\n",
      "3298\n",
      "3299\n",
      "3300\n",
      "3301\n",
      "3302\n",
      "3303\n",
      "3304\n",
      "3305\n",
      "3306\n",
      "3307\n",
      "3308\n",
      "3309\n",
      "3310\n",
      "3311\n",
      "3312\n",
      "3313\n",
      "3314\n",
      "3315\n",
      "3316\n",
      "3317\n",
      "3318\n",
      "3319\n",
      "3320\n",
      "3321\n",
      "3322\n",
      "3323\n",
      "3324\n",
      "3325\n",
      "3326\n",
      "3327\n",
      "3328\n",
      "130.0%/ 100% done\n",
      "3329\n",
      "3330\n",
      "3331\n",
      "3332\n",
      "3333\n",
      "3334\n",
      "3335\n",
      "3336\n",
      "3337\n",
      "3338\n",
      "3339\n",
      "3340\n",
      "3341\n",
      "3342\n",
      "3343\n",
      "3344\n",
      "3345\n",
      "3346\n",
      "3347\n",
      "3348\n",
      "3349\n",
      "3350\n",
      "3351\n",
      "3352\n",
      "3353\n",
      "3354\n",
      "3355\n",
      "3356\n",
      "3357\n",
      "3358\n",
      "3359\n",
      "3360\n",
      "3361\n",
      "3362\n",
      "3363\n",
      "3364\n",
      "3365\n",
      "3366\n",
      "3367\n",
      "3368\n",
      "3369\n",
      "3370\n",
      "3371\n",
      "3372\n",
      "3373\n",
      "3374\n",
      "3376\n",
      "3377\n",
      "3378\n",
      "3379\n",
      "3380\n",
      "3381\n",
      "3382\n",
      "3383\n",
      "3384\n",
      "3385\n",
      "3386\n",
      "3387\n",
      "3388\n",
      "3389\n",
      "3390\n",
      "3391\n",
      "3392\n",
      "3393\n",
      "3394\n",
      "3395\n",
      "3396\n",
      "3397\n",
      "3398\n",
      "3399\n",
      "3400\n",
      "3401\n",
      "3402\n",
      "3403\n",
      "3404\n",
      "3405\n",
      "3407\n",
      "3408\n",
      "3409\n",
      "3410\n",
      "3411\n",
      "3412\n",
      "3413\n",
      "3414\n",
      "3415\n",
      "3416\n",
      "3417\n",
      "3418\n",
      "3419\n",
      "3420\n",
      "3421\n",
      "3422\n",
      "3423\n",
      "3424\n",
      "3425\n",
      "3426\n",
      "3427\n",
      "3428\n",
      "3429\n",
      "3430\n",
      "3431\n",
      "3432\n",
      "3433\n",
      "3434\n",
      "3435\n",
      "3436\n",
      "3437\n",
      "3438\n",
      "3439\n",
      "3440\n",
      "3441\n",
      "3442\n",
      "3443\n",
      "3444\n",
      "3445\n",
      "3446\n",
      "3447\n",
      "3448\n",
      "3449\n",
      "3450\n",
      "3451\n",
      "3452\n",
      "3453\n",
      "3454\n",
      "3455\n",
      "3456\n",
      "3457\n",
      "3458\n",
      "3459\n",
      "3460\n",
      "3461\n",
      "3464\n",
      "3465\n",
      "3466\n",
      "3467\n",
      "3468\n",
      "3469\n",
      "3470\n",
      "3471\n",
      "3472\n",
      "3473\n",
      "3474\n",
      "3475\n",
      "3476\n",
      "3477\n",
      "3478\n",
      "3479\n",
      "3480\n",
      "3481\n",
      "3482\n",
      "3483\n",
      "3484\n",
      "3485\n",
      "3486\n",
      "3487\n",
      "3488\n",
      "3489\n",
      "3490\n",
      "3491\n",
      "3492\n",
      "3493\n",
      "3494\n",
      "3495\n",
      "3496\n",
      "3497\n",
      "3498\n",
      "3499\n",
      "3500\n",
      "3502\n",
      "3503\n",
      "3504\n",
      "3505\n",
      "3506\n",
      "3507\n",
      "3508\n",
      "3509\n",
      "3510\n",
      "3512\n",
      "3513\n",
      "3514\n",
      "3515\n",
      "3516\n",
      "3519\n",
      "3520\n",
      "3521\n",
      "3522\n",
      "3523\n",
      "3524\n",
      "3525\n",
      "3526\n",
      "3528\n",
      "3529\n",
      "3530\n",
      "3531\n",
      "3532\n",
      "3533\n",
      "3534\n",
      "3535\n",
      "3536\n",
      "3537\n",
      "3538\n",
      "3539\n",
      "3540\n",
      "3541\n",
      "3542\n",
      "3543\n",
      "3544\n",
      "3550\n",
      "3551\n",
      "3554\n",
      "3555\n",
      "3556\n",
      "3557\n",
      "3558\n",
      "3559\n",
      "3560\n",
      "3561\n",
      "3562\n",
      "3563\n",
      "3564\n",
      "3565\n",
      "3566\n",
      "3567\n",
      "3568\n",
      "3569\n",
      "3570\n",
      "3571\n",
      "3572\n",
      "3573\n",
      "3574\n",
      "3575\n",
      "3576\n",
      "3577\n",
      "3578\n",
      "3579\n",
      "3580\n",
      "3581\n",
      "3582\n",
      "3583\n",
      "3584\n",
      "140.0%/ 100% done\n",
      "3585\n",
      "3586\n",
      "3587\n",
      "3588\n",
      "3589\n",
      "3590\n",
      "3591\n",
      "3592\n",
      "3593\n",
      "3594\n",
      "3596\n",
      "3597\n",
      "3598\n",
      "3599\n",
      "3600\n",
      "3601\n",
      "3602\n",
      "3603\n",
      "3604\n",
      "3605\n",
      "3606\n",
      "3607\n",
      "3608\n",
      "3609\n",
      "3610\n",
      "3611\n",
      "3612\n",
      "3613\n",
      "3614\n",
      "3615\n",
      "3616\n",
      "3617\n",
      "3618\n",
      "3619\n",
      "3620\n",
      "3622\n",
      "3623\n",
      "3624\n",
      "3625\n",
      "3626\n",
      "3627\n",
      "3628\n",
      "3629\n",
      "3630\n",
      "3631\n",
      "3632\n",
      "3633\n",
      "3634\n",
      "3635\n",
      "3636\n",
      "3637\n",
      "3638\n",
      "3639\n",
      "3640\n",
      "3641\n",
      "3642\n",
      "3643\n",
      "3644\n",
      "3645\n",
      "3646\n",
      "3647\n",
      "3648\n",
      "3649\n",
      "3650\n",
      "3651\n",
      "3652\n",
      "3653\n",
      "3654\n",
      "3655\n",
      "3656\n",
      "3657\n",
      "3658\n",
      "3659\n",
      "3661\n",
      "3662\n",
      "3663\n",
      "3664\n",
      "3665\n",
      "3666\n",
      "3667\n",
      "3668\n",
      "3669\n",
      "3670\n",
      "3671\n",
      "3672\n",
      "3673\n",
      "3674\n",
      "3675\n",
      "3676\n",
      "3677\n",
      "3678\n",
      "3680\n",
      "3681\n",
      "3682\n",
      "3683\n",
      "3684\n",
      "3685\n",
      "3686\n",
      "3687\n",
      "3688\n",
      "3689\n",
      "3690\n",
      "3691\n",
      "3692\n",
      "3693\n",
      "3694\n",
      "3695\n",
      "3696\n",
      "3697\n",
      "3698\n",
      "3699\n",
      "3700\n",
      "3702\n",
      "3704\n",
      "3705\n",
      "3706\n",
      "3707\n",
      "3708\n",
      "3709\n",
      "3710\n",
      "3711\n",
      "3712\n",
      "3713\n",
      "3714\n",
      "3715\n",
      "3716\n",
      "3717\n",
      "3718\n",
      "3719\n",
      "3720\n",
      "3721\n",
      "3722\n",
      "3723\n",
      "3724\n",
      "3725\n",
      "3726\n",
      "3728\n",
      "3729\n",
      "3730\n",
      "3731\n",
      "3732\n",
      "3733\n",
      "3737\n",
      "3738\n",
      "3739\n",
      "3740\n",
      "3741\n",
      "3742\n",
      "3743\n",
      "3744\n",
      "3745\n",
      "3746\n",
      "3747\n",
      "3748\n",
      "3749\n",
      "3750\n",
      "3751\n",
      "3752\n",
      "3753\n",
      "3754\n",
      "3755\n",
      "3756\n",
      "3757\n",
      "3758\n",
      "3759\n",
      "3760\n",
      "3761\n",
      "3762\n",
      "3763\n",
      "3764\n",
      "3765\n",
      "3766\n",
      "3767\n",
      "3768\n",
      "3769\n",
      "3770\n",
      "3771\n",
      "3772\n",
      "3773\n",
      "3774\n",
      "3775\n",
      "3776\n",
      "3777\n",
      "3778\n",
      "3779\n",
      "3780\n",
      "3781\n",
      "3782\n",
      "3783\n",
      "3784\n",
      "3785\n",
      "3786\n",
      "3787\n",
      "3788\n",
      "3789\n",
      "3790\n",
      "3791\n",
      "3792\n",
      "3793\n",
      "3794\n",
      "3795\n",
      "3799\n",
      "3800\n",
      "3801\n",
      "3802\n",
      "3803\n",
      "3804\n",
      "3805\n",
      "3806\n",
      "3807\n",
      "3808\n",
      "3809\n",
      "3810\n",
      "3811\n",
      "3812\n",
      "3813\n",
      "3814\n",
      "3816\n",
      "3817\n",
      "3818\n",
      "3819\n",
      "3820\n",
      "3821\n",
      "3822\n",
      "3823\n",
      "3824\n",
      "3825\n",
      "3826\n",
      "3827\n",
      "3828\n",
      "3830\n",
      "3831\n",
      "3832\n",
      "3833\n",
      "3834\n",
      "3835\n",
      "3836\n",
      "3837\n",
      "3838\n",
      "3839\n",
      "3840\n",
      "150.0%/ 100% done\n",
      "3841\n",
      "3842\n",
      "3843\n",
      "3844\n",
      "3845\n",
      "3846\n",
      "3847\n",
      "3848\n",
      "3849\n",
      "3850\n",
      "3851\n",
      "3852\n",
      "3853\n",
      "3854\n",
      "3855\n",
      "3856\n",
      "3857\n",
      "3858\n",
      "3859\n",
      "3860\n",
      "3861\n",
      "3862\n",
      "3863\n",
      "3864\n",
      "3865\n",
      "3866\n",
      "3867\n",
      "3868\n",
      "3869\n",
      "3870\n",
      "3871\n",
      "3872\n",
      "3873\n",
      "3874\n",
      "3875\n",
      "3876\n",
      "3877\n",
      "3878\n",
      "3879\n",
      "3880\n",
      "3882\n",
      "3883\n",
      "3884\n",
      "3885\n",
      "3886\n",
      "3887\n",
      "3888\n",
      "3889\n",
      "3890\n",
      "3891\n",
      "3892\n",
      "3893\n",
      "3894\n",
      "3895\n",
      "3896\n",
      "3897\n",
      "3898\n",
      "3899\n",
      "3900\n",
      "3901\n",
      "3902\n",
      "3903\n",
      "3904\n",
      "3905\n",
      "3906\n",
      "3907\n",
      "3908\n",
      "3909\n",
      "3910\n",
      "3911\n",
      "3912\n",
      "3913\n",
      "3914\n",
      "3915\n",
      "3916\n",
      "3917\n",
      "3918\n",
      "3919\n",
      "3920\n",
      "3921\n",
      "3922\n",
      "3923\n",
      "3924\n",
      "3925\n",
      "3926\n",
      "3927\n",
      "3928\n",
      "3929\n",
      "3930\n",
      "3931\n",
      "3932\n",
      "3933\n",
      "3934\n",
      "3935\n",
      "3936\n",
      "3937\n",
      "3938\n",
      "3939\n",
      "3940\n",
      "3941\n",
      "3942\n",
      "3943\n",
      "3944\n",
      "3945\n",
      "3946\n",
      "3948\n",
      "3949\n",
      "3950\n",
      "3951\n",
      "3953\n",
      "3954\n",
      "3955\n",
      "3956\n",
      "3957\n",
      "3958\n",
      "3959\n",
      "3960\n",
      "3961\n",
      "3962\n"
     ]
    }
   ],
   "source": [
    "pd.options.mode.chained_assignment = None\n",
    "# group_sv_by_gene(merged_df.head(5000))\n",
    "in_df, out_df = group_sv_by_gene_single_cohort(sv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_grouped_sv_out.csv\", index=False)\n",
    "out_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_grouped_cluster_out.csv\", index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "breakdown of SV by type shared to AP for double checking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64,\n",
       " insertion               2648\n",
       " deletion                1234\n",
       " duplication_inverted      30\n",
       " duplication               30\n",
       " inversion_paired          13\n",
       " duplication_split          4\n",
       " inversion                  3\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_df['Type'].value_counts(),out_df['Type'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solve Clustering (Solve Cluster then filter)\n",
    "compare with Solve clustering results"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "extract header and rows, ignoring other # lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output file path: Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_minimal.smap.smap\n"
     ]
    }
   ],
   "source": [
    "in_path = \"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_cluster_molecule_variant.txt\"\n",
    "out_path = \"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_minimal.smap\"\n",
    "get_smap_table(in_path, out_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load csv as DF\n",
    "solve_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_minimal.smap\", sep='\\t', dtype=object,  index_col=False)\n",
    "solve_df = solve_df.assign(Sample_ID = \"GM24385\")\n",
    "solve_df.rename(columns={\"RefcontigID1\":\"chr\",\"SmapEntryID\":\"SmapID\"}, inplace=True)\n",
    "solve_df = solve_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'SVsize':'float64'})"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "variant clustering smmary was manually edited to remove '#' lines, keepingonly headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "insertion                 5827\n",
       "deletion                  3385\n",
       "inversion                   67\n",
       "translocation_intrachr      44\n",
       "translocation_interchr      32\n",
       "duplication_direct          30\n",
       "duplication_inverted        24\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_cluster_df = pd.read_csv(r\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_cluster_molecule_variant_summary.txt\", sep='\\t',dtype=object,  index_col=False)\n",
    "solve_cluster_df['type'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "flatten inversions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['insertion', 'deletion', 'duplication_inverted', 'duplication',\n",
       "       'duplication_split', 'inversion', 'inversion_partial',\n",
       "       'inversion_paired', 'inversion_repeat', 'translocation_intrachr',\n",
       "       'translocation_interchr', 'trans_intrachr_repeat'], dtype=object)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.unique(solve_df['Type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total: 16055\n",
      "inversions, inversion nbase, and inversion & inversion_partials: 388\n"
     ]
    }
   ],
   "source": [
    "solve_inv_flat_df = flatten_df_inv(solve_df)\n",
    "solve_inv_flat_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_inv_flat_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusterId</th>\n",
       "      <th>SmapID</th>\n",
       "      <th>QryContigID</th>\n",
       "      <th>chr</th>\n",
       "      <th>RefcontigID2</th>\n",
       "      <th>QryStartPos</th>\n",
       "      <th>QryEndPos</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>GenotypeGroup</th>\n",
       "      <th>RawConfidence</th>\n",
       "      <th>RawConfidenceLeft</th>\n",
       "      <th>RawConfidenceRight</th>\n",
       "      <th>RawConfidenceCenter</th>\n",
       "      <th>SVsize</th>\n",
       "      <th>SVfreq</th>\n",
       "      <th>orientation</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9267</td>\n",
       "      <td>148</td>\n",
       "      <td>502</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10389807.2</td>\n",
       "      <td>10543577.5</td>\n",
       "      <td>13217647.0</td>\n",
       "      <td>13242763.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>9.02</td>\n",
       "      <td>617.71</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>25116.5</td>\n",
       "      <td>0.438</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9268</td>\n",
       "      <td>146</td>\n",
       "      <td>501</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13976504.6</td>\n",
       "      <td>13990446.0</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>16942107.0</td>\n",
       "      <td>0.22</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>2098.24</td>\n",
       "      <td>22.53</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>325289.0</td>\n",
       "      <td>0.370</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9268</td>\n",
       "      <td>165</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>104099695.4</td>\n",
       "      <td>104113661.8</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>16942107.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>151.51</td>\n",
       "      <td>12.68</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>325289.0</td>\n",
       "      <td>0.203</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9268</td>\n",
       "      <td>173</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>104346895.3</td>\n",
       "      <td>104360861.8</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>16942107.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>153.26</td>\n",
       "      <td>14.27</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>325289.0</td>\n",
       "      <td>0.412</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9268</td>\n",
       "      <td>177</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>103480089.7</td>\n",
       "      <td>103499002.6</td>\n",
       "      <td>16566162.0</td>\n",
       "      <td>16859181.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>18.97</td>\n",
       "      <td>17794.52</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>293019.0</td>\n",
       "      <td>0.195</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  clusterId SmapID QryContigID  chr RefcontigID2  QryStartPos    QryEndPos  \\\n",
       "0      9267    148         502    1            1   10389807.2   10543577.5   \n",
       "1      9268    146         501    1            1   13976504.6   13990446.0   \n",
       "2      9268    165          41    1            1  104099695.4  104113661.8   \n",
       "3      9268    173          42    1            1  104346895.3  104360861.8   \n",
       "4      9268    177          41    1            1  103480089.7  103499002.6   \n",
       "\n",
       "   RefStartPos   RefEndPos Confidence  ... Genotype GenotypeGroup  \\\n",
       "0   13217647.0  13242763.5       0.00  ...        1            -1   \n",
       "1   16616818.0  16942107.0       0.22  ...        1            66   \n",
       "2   16616818.0  16942107.0       0.00  ...        2            66   \n",
       "3   16616818.0  16942107.0       0.00  ...        2            66   \n",
       "4   16566162.0  16859181.0       0.00  ...        2            66   \n",
       "\n",
       "  RawConfidence RawConfidenceLeft RawConfidenceRight RawConfidenceCenter  \\\n",
       "0         -1.00              9.02             617.71               -1.00   \n",
       "1         -1.00           2098.24              22.53               -1.00   \n",
       "2         -1.00            151.51              12.68               -1.00   \n",
       "3         -1.00            153.26              14.27               -1.00   \n",
       "4         -1.00             18.97           17794.52               -1.00   \n",
       "\n",
       "     SVsize SVfreq orientation Sample_ID  \n",
       "0   25116.5  0.438         NA    GM24385  \n",
       "1  325289.0  0.370         NA    GM24385  \n",
       "2  325289.0  0.203         NA    GM24385  \n",
       "3  325289.0  0.412         NA    GM24385  \n",
       "4  293019.0  0.195         NA    GM24385  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "((15861, 29),\n",
       " insertion                 9959\n",
       " deletion                  5388\n",
       " inversion                  172\n",
       " translocation_intrachr     123\n",
       " translocation_interchr      66\n",
       " duplication                 59\n",
       " duplication_inverted        53\n",
       " inversion_paired            21\n",
       " trans_intrachr_repeat       13\n",
       " duplication_split            6\n",
       " inversion_repeat             1\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(solve_inv_flat_df.head())\n",
    "solve_inv_flat_df.shape, solve_inv_flat_df['Type'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "keep non-translocations and exclude gains/loss, *masked, *common, *tiny, *nbase and complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Filter rare SVs and output selectd columns\n",
    "# df_rareSV = filter_rare_svs(inv_flat_df, 1.0)\n",
    "# # # df_rareSV = filter_rare_svs(case_df, 1.0)\n",
    "# n_rare = len(df_rareSV) \n",
    "\n",
    "# exclude unwanted sv types\n",
    "solve_df_excluded = exclude_types_svs(solve_inv_flat_df, True)\n",
    "# df_excluded = exclude_types_svs(inv_flat_df)\n",
    "# df_excluded = exclude_types_svs(case_df)\n",
    "n_exc = len(solve_df_excluded)\n",
    "\n",
    "# filter for confidence intervals\n",
    "solve_df_sig = filter_sig_svs(solve_df_excluded)\n",
    "n_sig = len(solve_df_sig)\n",
    "\n",
    "solve_df_sig.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_filtered_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusterId</th>\n",
       "      <th>SmapID</th>\n",
       "      <th>QryContigID</th>\n",
       "      <th>chr</th>\n",
       "      <th>RefcontigID2</th>\n",
       "      <th>QryStartPos</th>\n",
       "      <th>QryEndPos</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>GenotypeGroup</th>\n",
       "      <th>RawConfidence</th>\n",
       "      <th>RawConfidenceLeft</th>\n",
       "      <th>RawConfidenceRight</th>\n",
       "      <th>RawConfidenceCenter</th>\n",
       "      <th>SVsize</th>\n",
       "      <th>SVfreq</th>\n",
       "      <th>orientation</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>9270</td>\n",
       "      <td>707</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16633490.1</td>\n",
       "      <td>16633490.1</td>\n",
       "      <td>103711691.0</td>\n",
       "      <td>103734414.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>16069.91</td>\n",
       "      <td>5.25</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>22723.0</td>\n",
       "      <td>0.455</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>9273</td>\n",
       "      <td>959</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>531135.5</td>\n",
       "      <td>531135.5</td>\n",
       "      <td>149840746.0</td>\n",
       "      <td>149855154.0</td>\n",
       "      <td>0.98</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>48.29</td>\n",
       "      <td>2.72</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>14408.0</td>\n",
       "      <td>0.499</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>9278</td>\n",
       "      <td>2795</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>30390774.8</td>\n",
       "      <td>30389965.3</td>\n",
       "      <td>211826848.0</td>\n",
       "      <td>211840496.0</td>\n",
       "      <td>0.90</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>8376.11</td>\n",
       "      <td>4.80</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>13648.0</td>\n",
       "      <td>0.556</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>9281</td>\n",
       "      <td>7183</td>\n",
       "      <td>1860</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>249179.9</td>\n",
       "      <td>249179.9</td>\n",
       "      <td>5986390.0</td>\n",
       "      <td>6738295.0</td>\n",
       "      <td>0.86</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2710</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>61.41</td>\n",
       "      <td>33.66</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>751905.0</td>\n",
       "      <td>0.189</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>9281</td>\n",
       "      <td>7185</td>\n",
       "      <td>2402</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>218808.3</td>\n",
       "      <td>218808.3</td>\n",
       "      <td>5986390.0</td>\n",
       "      <td>6738295.0</td>\n",
       "      <td>0.86</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2710</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>31.70</td>\n",
       "      <td>49.95</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>751905.0</td>\n",
       "      <td>0.192</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   clusterId SmapID QryContigID  chr RefcontigID2 QryStartPos   QryEndPos  \\\n",
       "18      9270    707          42    1            1  16633490.1  16633490.1   \n",
       "25      9273    959         211    1            1    531135.5    531135.5   \n",
       "35      9278   2795          12    2            2  30390774.8  30389965.3   \n",
       "40      9281   7183        1860    7            7    249179.9    249179.9   \n",
       "41      9281   7185        2402    7            7    218808.3    218808.3   \n",
       "\n",
       "    RefStartPos    RefEndPos  Confidence  ... Genotype GenotypeGroup  \\\n",
       "18  103711691.0  103734414.0        0.99  ...        1            -1   \n",
       "25  149840746.0  149855154.0        0.98  ...        1            -1   \n",
       "35  211826848.0  211840496.0        0.90  ...        1            -1   \n",
       "40    5986390.0    6738295.0        0.86  ...        1          2710   \n",
       "41    5986390.0    6738295.0        0.86  ...        2          2710   \n",
       "\n",
       "   RawConfidence RawConfidenceLeft RawConfidenceRight RawConfidenceCenter  \\\n",
       "18         -1.00          16069.91               5.25               -1.00   \n",
       "25         -1.00             48.29               2.72               -1.00   \n",
       "35         -1.00           8376.11               4.80               -1.00   \n",
       "40         -1.00             61.41              33.66               -1.00   \n",
       "41         -1.00             31.70              49.95               -1.00   \n",
       "\n",
       "      SVsize SVfreq orientation Sample_ID  \n",
       "18   22723.0  0.455         NA    GM24385  \n",
       "25   14408.0  0.499         NA    GM24385  \n",
       "35   13648.0  0.556         NA    GM24385  \n",
       "40  751905.0  0.189         NA    GM24385  \n",
       "41  751905.0  0.192         NA    GM24385  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "((6467, 29),)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(solve_df_sig.head())\n",
    "solve_df_sig.shape,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusterId</th>\n",
       "      <th>SmapID</th>\n",
       "      <th>QryContigID</th>\n",
       "      <th>chr</th>\n",
       "      <th>RefcontigID2</th>\n",
       "      <th>QryStartPos</th>\n",
       "      <th>QryEndPos</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>GenotypeGroup</th>\n",
       "      <th>RawConfidence</th>\n",
       "      <th>RawConfidenceLeft</th>\n",
       "      <th>RawConfidenceRight</th>\n",
       "      <th>RawConfidenceCenter</th>\n",
       "      <th>SVsize</th>\n",
       "      <th>SVfreq</th>\n",
       "      <th>orientation</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9270</td>\n",
       "      <td>707</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16633490.1</td>\n",
       "      <td>16633490.1</td>\n",
       "      <td>103711691.0</td>\n",
       "      <td>103734414.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>16069.91</td>\n",
       "      <td>5.25</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>22723.0</td>\n",
       "      <td>0.455</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9273</td>\n",
       "      <td>959</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>531135.5</td>\n",
       "      <td>531135.5</td>\n",
       "      <td>149840746.0</td>\n",
       "      <td>149855154.0</td>\n",
       "      <td>0.98</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>48.29</td>\n",
       "      <td>2.72</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>14408.0</td>\n",
       "      <td>0.499</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9278</td>\n",
       "      <td>2795</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>30390774.8</td>\n",
       "      <td>30389965.3</td>\n",
       "      <td>211826848.0</td>\n",
       "      <td>211840496.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>8376.11</td>\n",
       "      <td>4.80</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>13648.0</td>\n",
       "      <td>0.556</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9281</td>\n",
       "      <td>7183</td>\n",
       "      <td>1860</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>249179.9</td>\n",
       "      <td>249179.9</td>\n",
       "      <td>5986390.0</td>\n",
       "      <td>6738295.0</td>\n",
       "      <td>0.86</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2710</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>61.41</td>\n",
       "      <td>33.66</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>751905.0</td>\n",
       "      <td>0.189</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9281</td>\n",
       "      <td>7185</td>\n",
       "      <td>2402</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>218808.3</td>\n",
       "      <td>218808.3</td>\n",
       "      <td>5986390.0</td>\n",
       "      <td>6738295.0</td>\n",
       "      <td>0.86</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2710</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>31.70</td>\n",
       "      <td>49.95</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>751905.0</td>\n",
       "      <td>0.192</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  clusterId SmapID QryContigID  chr RefcontigID2 QryStartPos   QryEndPos  \\\n",
       "0      9270    707          42    1            1  16633490.1  16633490.1   \n",
       "1      9273    959         211    1            1    531135.5    531135.5   \n",
       "2      9278   2795          12    2            2  30390774.8  30389965.3   \n",
       "3      9281   7183        1860    7            7    249179.9    249179.9   \n",
       "4      9281   7185        2402    7            7    218808.3    218808.3   \n",
       "\n",
       "   RefStartPos    RefEndPos Confidence  ... Genotype GenotypeGroup  \\\n",
       "0  103711691.0  103734414.0       0.99  ...        1            -1   \n",
       "1  149840746.0  149855154.0       0.98  ...        1            -1   \n",
       "2  211826848.0  211840496.0        0.9  ...        1            -1   \n",
       "3    5986390.0    6738295.0       0.86  ...        1          2710   \n",
       "4    5986390.0    6738295.0       0.86  ...        2          2710   \n",
       "\n",
       "  RawConfidence RawConfidenceLeft RawConfidenceRight RawConfidenceCenter  \\\n",
       "0         -1.00          16069.91               5.25               -1.00   \n",
       "1         -1.00             48.29               2.72               -1.00   \n",
       "2         -1.00           8376.11               4.80               -1.00   \n",
       "3         -1.00             61.41              33.66               -1.00   \n",
       "4         -1.00             31.70              49.95               -1.00   \n",
       "\n",
       "     SVsize SVfreq orientation Sample_ID  \n",
       "0   22723.0  0.455         NA    GM24385  \n",
       "1   14408.0  0.499         NA    GM24385  \n",
       "2   13648.0  0.556         NA    GM24385  \n",
       "3  751905.0  0.189         NA    GM24385  \n",
       "4  751905.0  0.192         NA    GM24385  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load compiled sv list\n",
    "solve_sig_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_filtered_out.csv\", dtype=object,  index_col=False)\n",
    "\n",
    "solve_sig_df = solve_sig_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "display(solve_sig_df.head())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ngaps and overlapgenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "234 has gaps, and 6233 has no gaps, for a total of 6467 SVs\n"
     ]
    }
   ],
   "source": [
    "# remove indel_ngaps\n",
    "# True if overlaps gaps, false if not, remove Falses\n",
    "solve_sig_df[\"has_gaps\"] = solve_sig_df.apply(lambda x: filter_nogaps(x[\"Type\"], x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "solve_nogaps_df = solve_sig_df.loc[solve_sig_df[\"has_gaps\"] == False]\n",
    "\n",
    "num_true = len(solve_sig_df[solve_sig_df[\"has_gaps\"] == True])\n",
    "num_false = len(solve_sig_df[solve_sig_df[\"has_gaps\"] == False])\n",
    "\n",
    "print(f\"{num_true} has gaps, and {num_false} has no gaps, for a total of {num_true + num_false} SVs\")\n",
    "\n",
    "# out_file(nogaps_df, \"case_cleaned_out\")\n",
    "solve_nogaps_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_nogaps_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "From 6233 cleaned SVs, 2202 SVs do not overlap any gene, while 4031 do.\n"
     ]
    }
   ],
   "source": [
    "# find overlap genes matching hg38knownCanonical.bed\n",
    "solve_nogaps_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_nogaps_out.csv\", dtype=object,  index_col=False)\n",
    "solve_nogaps_df = solve_nogaps_df.astype({'RefStartPos':'float64','RefEndPos':'float64','chr':'int'})\n",
    "solve_nogaps_df.dtypes\n",
    "\n",
    "solve_nogaps_df[\"OverlapGenes\"] = solve_nogaps_df.apply(lambda x: filter_genes(bed_df, x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "\n",
    "no_overlaps = len(solve_nogaps_df.loc[solve_nogaps_df[\"OverlapGenes\"] == \"-\"])\n",
    "overlaps = len(solve_nogaps_df.loc[solve_nogaps_df[\"OverlapGenes\"] != \"-\"])\n",
    "print(f\"From {no_overlaps + overlaps} cleaned SVs, {no_overlaps} SVs do not overlap any gene, while {overlaps} do.\")\n",
    "\n",
    "# out_file(nogaps_df, \"control_cleaned_out\")\n",
    "solve_nogaps_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_cleaned_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "solve_sv_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_cleaned_out.csv\", dtype=object,  index_col=False)\n",
    "solve_sv_df = solve_sv_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'SVsize':'float64', 'chr':'int'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "solve_sv_df['OverlapGenes'] = solve_sv_df.apply(lambda x: parse_sorted_unique_genes(x['OverlapGenes']), axis=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many filtered clusters?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64,\n",
       " insertion               2597\n",
       " deletion                1217\n",
       " duplication_direct        30\n",
       " duplication_inverted      24\n",
       " inversion                 16\n",
       " Name: type, dtype: int64)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_sv_df['Type'].value_counts(), solve_cluster_df.loc[solve_cluster_df['clusterId'].isin(solve_sv_df['clusterId'])]['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3884"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(solve_sv_df['clusterId'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['clusterId', 'SmapID', 'QryContigID', 'chr', 'RefcontigID2',\n",
       "       'QryStartPos', 'QryEndPos', 'RefStartPos', 'RefEndPos', 'Confidence',\n",
       "       'Type', 'XmapID1', 'XmapID2', 'LinkID', 'QryStartIdx', 'QryEndIdx',\n",
       "       'RefStartIdx', 'RefEndIdx', 'Zygosity', 'Genotype', 'GenotypeGroup',\n",
       "       'RawConfidence', 'RawConfidenceLeft', 'RawConfidenceRight',\n",
       "       'RawConfidenceCenter', 'SVsize', 'SVfreq', 'orientation', 'Sample_ID',\n",
       "       'has_gaps', 'OverlapGenes'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_sv_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "39\n",
      "60\n",
      "101\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[28], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m pd\u001b[39m.\u001b[39moptions\u001b[39m.\u001b[39mmode\u001b[39m.\u001b[39mchained_assignment \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[39m# group_sv_by_gene(merged_df.head(5000))\u001b[39;00m\n\u001b[1;32m----> 3\u001b[0m solve_in_df, solve_out_df \u001b[39m=\u001b[39m group_sv_by_gene_single_cohort(solve_sv_df)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\enrichment_util.py:512\u001b[0m, in \u001b[0;36mgroup_sv_by_gene_single_cohort\u001b[1;34m(cleaned_df)\u001b[0m\n\u001b[0;32m    509\u001b[0m \u001b[39m# main loop to process every SV group into smaller clusters, appends clustered dataframe into out_df\u001b[39;00m\n\u001b[0;32m    510\u001b[0m \u001b[39mfor\u001b[39;00m df \u001b[39min\u001b[39;00m groups\u001b[39m.\u001b[39mvalues():\n\u001b[0;32m    511\u001b[0m     \u001b[39m# inputs certain columns for clustering\u001b[39;00m\n\u001b[1;32m--> 512\u001b[0m     prev_cluster_ID, cluster_df_in, cluster_df_out \u001b[39m=\u001b[39m cluster_refpos_single(prev_cluster_ID, df, \u001b[39m5000.0\u001b[39;49m, \u001b[39m50\u001b[39;49m)\n\u001b[0;32m    514\u001b[0m     \u001b[39m# returned cluster SVs are appended to output df.\u001b[39;00m\n\u001b[0;32m    515\u001b[0m     df_in \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mconcat([df_in, cluster_df_in], axis\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m, ignore_index\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\enrichment_util.py:486\u001b[0m, in \u001b[0;36mcluster_refpos_single\u001b[1;34m(prev_cluster_ID, df_in, posWindow, reciprocalSize)\u001b[0m\n\u001b[0;32m    477\u001b[0m     \u001b[39m# row['num_overlap_DGV_calls'] = \", \".join(list(pd.unique(cluster_df['num_overlap_DGV_calls'].dropna())))\u001b[39;00m\n\u001b[0;32m    478\u001b[0m     \u001b[39m# print(row)\u001b[39;00m\n\u001b[0;32m    479\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    482\u001b[0m     \u001b[39m# cluster_df = cluster_df.drop_duplicates(subset=['Sample_ID'])\u001b[39;00m\n\u001b[0;32m    483\u001b[0m     \u001b[39m# display(cluster_df)\u001b[39;00m\n\u001b[0;32m    485\u001b[0m     sv_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mconcat([cluster_df, sv_df], axis\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m, ignore_index\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m--> 486\u001b[0m     row_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mconcat([row_df, row], axis\u001b[39m=\u001b[39;49m\u001b[39m0\u001b[39;49m, ignore_index\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)\n\u001b[0;32m    488\u001b[0m \u001b[39mreturn\u001b[39;00m prev_cluster_ID, sv_df, row_df\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\enrichment\\Lib\\site-packages\\pandas\\util\\_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    325\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(args) \u001b[39m>\u001b[39m num_allow_args:\n\u001b[0;32m    326\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[0;32m    327\u001b[0m         msg\u001b[39m.\u001b[39mformat(arguments\u001b[39m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[0;32m    328\u001b[0m         \u001b[39mFutureWarning\u001b[39;00m,\n\u001b[0;32m    329\u001b[0m         stacklevel\u001b[39m=\u001b[39mfind_stack_level(),\n\u001b[0;32m    330\u001b[0m     )\n\u001b[1;32m--> 331\u001b[0m \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\enrichment\\Lib\\site-packages\\pandas\\core\\reshape\\concat.py:381\u001b[0m, in \u001b[0;36mconcat\u001b[1;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[0;32m    159\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    160\u001b[0m \u001b[39mConcatenate pandas objects along a particular axis.\u001b[39;00m\n\u001b[0;32m    161\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    366\u001b[0m \u001b[39m1   3   4\u001b[39;00m\n\u001b[0;32m    367\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    368\u001b[0m op \u001b[39m=\u001b[39m _Concatenator(\n\u001b[0;32m    369\u001b[0m     objs,\n\u001b[0;32m    370\u001b[0m     axis\u001b[39m=\u001b[39maxis,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    378\u001b[0m     sort\u001b[39m=\u001b[39msort,\n\u001b[0;32m    379\u001b[0m )\n\u001b[1;32m--> 381\u001b[0m \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mget_result()\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\enrichment\\Lib\\site-packages\\pandas\\core\\reshape\\concat.py:616\u001b[0m, in \u001b[0;36m_Concatenator.get_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    612\u001b[0m             indexers[ax] \u001b[39m=\u001b[39m obj_labels\u001b[39m.\u001b[39mget_indexer(new_labels)\n\u001b[0;32m    614\u001b[0m     mgrs_indexers\u001b[39m.\u001b[39mappend((obj\u001b[39m.\u001b[39m_mgr, indexers))\n\u001b[1;32m--> 616\u001b[0m new_data \u001b[39m=\u001b[39m concatenate_managers(\n\u001b[0;32m    617\u001b[0m     mgrs_indexers, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mnew_axes, concat_axis\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mbm_axis, copy\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcopy\n\u001b[0;32m    618\u001b[0m )\n\u001b[0;32m    619\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcopy:\n\u001b[0;32m    620\u001b[0m     new_data\u001b[39m.\u001b[39m_consolidate_inplace()\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\enrichment\\Lib\\site-packages\\pandas\\core\\internals\\concat.py:233\u001b[0m, in \u001b[0;36mconcatenate_managers\u001b[1;34m(mgrs_indexers, axes, concat_axis, copy)\u001b[0m\n\u001b[0;32m    231\u001b[0m     fastpath \u001b[39m=\u001b[39m blk\u001b[39m.\u001b[39mvalues\u001b[39m.\u001b[39mdtype \u001b[39m==\u001b[39m values\u001b[39m.\u001b[39mdtype\n\u001b[0;32m    232\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 233\u001b[0m     values \u001b[39m=\u001b[39m _concatenate_join_units(join_units, concat_axis, copy\u001b[39m=\u001b[39;49mcopy)\n\u001b[0;32m    234\u001b[0m     fastpath \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m\n\u001b[0;32m    236\u001b[0m \u001b[39mif\u001b[39;00m fastpath:\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\enrichment\\Lib\\site-packages\\pandas\\core\\internals\\concat.py:577\u001b[0m, in \u001b[0;36m_concatenate_join_units\u001b[1;34m(join_units, concat_axis, copy)\u001b[0m\n\u001b[0;32m    574\u001b[0m     concat_values \u001b[39m=\u001b[39m ensure_block_shape(concat_values, \u001b[39m2\u001b[39m)\n\u001b[0;32m    576\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 577\u001b[0m     concat_values \u001b[39m=\u001b[39m concat_compat(to_concat, axis\u001b[39m=\u001b[39;49mconcat_axis)\n\u001b[0;32m    579\u001b[0m \u001b[39mreturn\u001b[39;00m concat_values\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\enrichment\\Lib\\site-packages\\pandas\\core\\dtypes\\concat.py:151\u001b[0m, in \u001b[0;36mconcat_compat\u001b[1;34m(to_concat, axis, ea_compat_axis)\u001b[0m\n\u001b[0;32m    148\u001b[0m             to_concat \u001b[39m=\u001b[39m [x\u001b[39m.\u001b[39mastype(\u001b[39m\"\u001b[39m\u001b[39mobject\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mfor\u001b[39;00m x \u001b[39min\u001b[39;00m to_concat]\n\u001b[0;32m    149\u001b[0m             kinds \u001b[39m=\u001b[39m {\u001b[39m\"\u001b[39m\u001b[39mo\u001b[39m\u001b[39m\"\u001b[39m}\n\u001b[1;32m--> 151\u001b[0m result \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39;49mconcatenate(to_concat, axis\u001b[39m=\u001b[39;49maxis)\n\u001b[0;32m    152\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m kinds \u001b[39mand\u001b[39;00m result\u001b[39m.\u001b[39mdtype\u001b[39m.\u001b[39mkind \u001b[39min\u001b[39;00m [\u001b[39m\"\u001b[39m\u001b[39mi\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mu\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m]:\n\u001b[0;32m    153\u001b[0m     \u001b[39m# GH#39817\u001b[39;00m\n\u001b[0;32m    154\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[0;32m    155\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mBehavior when concatenating bool-dtype and numeric-dtype arrays is \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    156\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mdeprecated; in a future version these will cast to object dtype \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    160\u001b[0m         stacklevel\u001b[39m=\u001b[39mfind_stack_level(),\n\u001b[0;32m    161\u001b[0m     )\n",
      "File \u001b[1;32m<__array_function__ internals>:200\u001b[0m, in \u001b[0;36mconcatenate\u001b[1;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "pd.options.mode.chained_assignment = None\n",
    "# group_sv_by_gene(merged_df.head(5000))\n",
    "solve_in_df, solve_out_df = group_sv_by_gene_single_cohort(solve_sv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "solve_in_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_grouped_sv_out.csv\", index=False)\n",
    "solve_out_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_cluster_grouped_cluster_out.csv\", index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "breakdown of SV by type shared to AP for double checking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64,\n",
       " insertion               2648\n",
       " deletion                1234\n",
       " duplication_inverted      30\n",
       " duplication               30\n",
       " inversion_paired          13\n",
       " duplication_split          4\n",
       " inversion                  3\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_in_df['Type'].value_counts(),solve_out_df['Type'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "also inspect variant summary file"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### inspect clusters between solve and cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "solve_cID = solve_in_df['clusterId'].unique()\n",
    "# cohort_cID = solve_in_df['cluster_ID']\n",
    "\n",
    "discordant = []\n",
    "for i in solve_cID:\n",
    "    IDs = solve_in_df.loc[solve_in_df['clusterId'] == i]['cluster_ID']\n",
    "    if len(set(IDs)) > 1:\n",
    "        discordant.append(i)\n",
    "\n",
    "len(discordant)\n",
    "\n",
    "# mark rows with discordant cluster IDs\n",
    "solve_in_df['discordant'] = solve_in_df.apply(lambda x: True if x['clusterId'] in (discordant) else False, axis=1)\n",
    "solve_in_df.loc[solve_in_df['discordant'] == True].to_excel(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_discordant_cluster.xlsx\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solve Clustering 2.0: (Filter then Solve Cluster)\n",
    "`Output\\output_hg002_clustering_04Jul2023\\GM24385_case_cleaned_out.csv` converted to .smap and used as input for clustering. Output file treated as `solve_in_df`, using `cluster_ID` column from Solve variant_clustering binary.\n",
    "\n",
    "Path to .smap: `Enrichment Project\\clustering_optimization_enrichment\\Output\\output_hg002_clustering_04Jul2023\\filter_then_solve_clustering\\GM24385_case_cleaned_out.smap`"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "load filtered .csv and transform to .smap (\"#\" headers manually added)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_df = pd.read_csv(r\"Output\\output_hg002_clustering_04Jul2023\\filter_then_solve_clustering\\GM24385_case_cleaned_out.csv\", dtype=object,  index_col=False)\n",
    "sv_df.to_csv(r\"Output\\output_hg002_clustering_04Jul2023\\filter_then_solve_clustering\\GM24385_case_cleaned_out.smap\", sep ='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "clustering done using solve arguments"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "convert to header + table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# in_path = r\"Output\\output_hg002_clustering_04Jul2023\\filter_then_solve_clustering\\GM24385_solve_cluster_molecule_variant_summary.txt\"\n",
    "# out_path = r\"Output\\output_hg002_clustering_04Jul2023\\filter_then_solve_clustering\\GM24385_solve_clusters.txt\"\n",
    "# get_smap_table(in_path, out_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_sv_df = pd.read_csv(r\"Output\\output_hg002_clustering_04Jul2023\\filter_then_solve_clustering\\GM24385_solve_cluster_molecule_variant.txt\", sep='\\t',dtype=object,  index_col=False)\n",
    "final_cluster_df = pd.read_csv(r\"Output\\output_hg002_clustering_04Jul2023\\filter_then_solve_clustering\\GM24385_solve_cluster_molecule_variant_summary.txt\", sep='\\t',dtype=object,  index_col=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64,\n",
       " insertion               2597\n",
       " deletion                1217\n",
       " duplication_direct        30\n",
       " duplication_inverted      24\n",
       " Name: type, dtype: int64)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_sv_df['Type'].value_counts(),final_cluster_df['type'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment Single Clustering (filter then cluster sans OverlapGenes)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Discrepancies might be due to clustering parameters\n",
    "\n",
    "enrichment clustering accounts for OverlapGenes, while solve clustering doesn't"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_df = pd.read_csv(\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_cleaned_out.csv\", dtype=object,  index_col=False)\n",
    "cleaned_df = cleaned_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'SVsize':'float64', 'chr':'int'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_df['OverlapGenes'] = cleaned_df.apply(lambda x: parse_sorted_unique_genes(x['OverlapGenes']), axis=1)\n",
    "# cleaned_df.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\GM24385_case_cleaned_out.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((6233, 30),\n",
       " insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_df.shape, cleaned_df['Type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Single cohort clustering\n",
    "# 2 Helper functions to cluster SVs. Clustering removes redundant SVs, keeping only `unique` SVs\n",
    "pd.options.mode.chained_assignment = None\n",
    "# Helper function to cluster gene coordinates, given a group of SV(s) grouped by gene, chr, zygo, and type\n",
    "def cluster_refpos_single_nogenes(prev_cluster_ID, df_in, posWindow, reciprocalSize):\n",
    "    \n",
    "    # Get sv info from first df item\n",
    "    sv_info_headers = [\"OverlapGenes\",\"Zygosity\",\"chr\", \"Type\"]\n",
    "    sv_info = df_in[sv_info_headers].head(1)\n",
    "    # display(df_in.head(1))\n",
    "\n",
    "    qry_type = df_in[\"Type\"].iat[0]\n",
    "\n",
    "    sv_df = pd.DataFrame()\n",
    "    row_df = pd.DataFrame()\n",
    "    while False in pd.unique(df_in['clustered']):\n",
    "        \n",
    "        # subset non-clustered SVs and get first item\n",
    "        # df_in = df_in.loc[df_in['clustered'] == False, ['RefStartPos', 'RefEndPos', 'SVsize', 'case_ID', 'ctrl_ID', 'num_overlap_DGV_calls', 'clustered']]\n",
    "        df_in = df_in.loc[df_in['clustered'] == False]\n",
    "        \n",
    "        # get gene coordinates and size of first item\n",
    "        qry_start = float(df_in['RefStartPos'].iat[0]) + posWindow\n",
    "        qry_end = float(df_in['RefEndPos'].iat[0]) + posWindow\n",
    "        qry_size = float(df_in['SVsize'].iat[0])\n",
    "\n",
    "        # find overlaps with first item (special treatnment for inversions since no inv size in ctrldb)\n",
    "        if qry_type in ['ins','insertion', 'del' ,'deletion', 'dup','duplication_paired' ,'duplication_inverted', 'duplication', 'duplication_split']:\n",
    "        \n",
    "            cluster_df = df_in.loc[(((qry_start - posWindow <= df_in['RefStartPos']) & (df_in['RefStartPos'] <= qry_end + posWindow)) \\\n",
    "                                    | ((df_in['RefStartPos'] <= qry_start - posWindow) & (qry_start - posWindow <= df_in['RefEndPos']))) \\\n",
    "                                    & (((qry_size/df_in['SVsize'])*100 >= reciprocalSize) \\\n",
    "                                        & ((df_in['SVsize']/qry_size)*100 >= reciprocalSize))]\n",
    "    \n",
    "            \n",
    "            # mark as clustered in original input_df\n",
    "            df_in.loc[(((qry_start - posWindow <= df_in['RefStartPos']) & (df_in['RefStartPos'] <= qry_end + posWindow)) \\\n",
    "                                    | ((df_in['RefStartPos'] <= qry_start - posWindow) & (qry_start - posWindow <= df_in['RefEndPos']))) \\\n",
    "                                    & (((qry_size/df_in['SVsize'])*100 >= reciprocalSize) \\\n",
    "                                        & ((df_in['SVsize']/qry_size)*100 >= reciprocalSize)), 'clustered'] = True\n",
    "\n",
    "        else: # else if its inv (or trans)\n",
    "\n",
    "            cluster_df = df_in.loc[(((qry_start - posWindow <= df_in['RefStartPos']) & (df_in['RefStartPos'] <= qry_end + posWindow)) \\\n",
    "                                    | ((df_in['RefStartPos'] <= qry_start - posWindow) & (qry_start - posWindow <= df_in['RefEndPos'])))]\n",
    "            \n",
    "            df_in.loc[(((qry_start - posWindow <= df_in['RefStartPos']) & (df_in['RefStartPos'] <= qry_end + posWindow)) \\\n",
    "                                    | ((df_in['RefStartPos'] <= qry_start - posWindow) & (qry_start - posWindow <= df_in['RefEndPos']))), 'clustered'] = True\n",
    "\n",
    "\n",
    "        # get cluster info\n",
    "        # get SV key information (OverlapGenes, Chr, Type, Zygo)\n",
    "        row = sv_info\n",
    "        \n",
    "        # set cluster and row ID number\n",
    "        prev_cluster_ID += 1\n",
    "        cluster_df['cluster_ID'] = prev_cluster_ID\n",
    "        row['cluster_ID'] = prev_cluster_ID\n",
    "\n",
    "        # get list of gene coordinates \n",
    "        row['RefStartPos'] = cluster_df['RefStartPos'].min()\n",
    "        row['RefEndPos'] = cluster_df['RefEndPos'].max()\n",
    "        row['listRefStartPos'] = \", \".join(cluster_df['RefStartPos'].astype(str))\n",
    "        row['listRefEndPos'] = \", \".join(cluster_df['RefEndPos'].astype(str))\n",
    "\n",
    "        # Get info on n samples that share an SV, including and excluding repeats\n",
    "        row['num_SVs'] = len(cluster_df)\n",
    "        row['num_unique_samples'] = len(cluster_df['Sample_ID'].dropna().unique())\n",
    "\n",
    "        row['Sample_ID'] =  \", \".join(cluster_df['Sample_ID'])\n",
    "\n",
    "        # populate lists of other sample-specific metrics for table QC\n",
    "\n",
    "        row['SVsize'] = \", \".join(cluster_df['SVsize'].astype(str))\n",
    "        row['sv_means'] = cluster_df['SVsize'].mean()\n",
    "        row['sv_std'] = cluster_df['SVsize'].std()\n",
    "        # row['num_overlap_DGV_calls'] = \", \".join(list(pd.unique(cluster_df['num_overlap_DGV_calls'].dropna())))\n",
    "        # print(row)\n",
    "\n",
    "        # display(cluster_df)\n",
    "        # Remove duplicate sample_ID and keep only 1 SV in a cluster per sample\n",
    "        # cluster_df = cluster_df.drop_duplicates(subset=['Sample_ID'])\n",
    "        cluster_df['clustered'] = True\n",
    "        \n",
    "        sv_df = pd.concat([cluster_df, sv_df], axis=0, ignore_index=True)\n",
    "        row_df = pd.concat([row_df, row], axis=0, ignore_index=True)\n",
    "        \n",
    "    return prev_cluster_ID, sv_df, row_df\n",
    "\n",
    "\n",
    "def cluster_sv_single_cohort_nogenes(cleaned_df):\n",
    "    # initialize column of T/F, where SVs == T have been clustered, and F haven't.\n",
    "    cleaned_df.loc[:, 'clustered'] = False    \n",
    "    cleaned_df.loc[:, 'cluster_ID'] = -1\n",
    "\n",
    "    # Sort first by refstartstop from smallest to largest\n",
    "    cleaned_df = cleaned_df.sort_values([\"chr\", \"Type\", 'RefStartPos','RefEndPos', 'SVsize'])\n",
    "\n",
    "    # Group compiled rare SVs by criteria, then list samples sharing each rare SV.\n",
    "    # due to memory issues, we put df_g in a dictionary, using identified rare SVs as key, then iterate by key to values containing df\n",
    "    groups = dict(list(cleaned_df.groupby(by=[\"chr\", \"Type\"])))\n",
    "\n",
    "    df_in = pd.DataFrame()\n",
    "    df_out = pd.DataFrame()\n",
    "\n",
    "    total = len(groups.values())\n",
    "    prev_cluster_ID = 0\n",
    "\n",
    "    # main loop to process every SV group into smaller clusters, appends clustered dataframe into out_df\n",
    "    for df in groups.values():\n",
    "        # inputs certain columns for clustering\n",
    "        prev_cluster_ID, cluster_df_in, cluster_df_out = cluster_refpos_single_nogenes(prev_cluster_ID, df, 5000.0, 50)\n",
    "\n",
    "        # returned cluster SVs are appended to output df.\n",
    "        df_in = pd.concat([df_in, cluster_df_in], axis=0, ignore_index=True)\n",
    "        df_out = pd.concat([df_out, cluster_df_out], axis=0, ignore_index=True)\n",
    "\n",
    "        print(prev_cluster_ID)\n",
    "\n",
    "    return df_in, df_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "97\n",
      "98\n",
      "102\n",
      "303\n",
      "305\n",
      "406\n",
      "408\n",
      "569\n",
      "570\n",
      "642\n",
      "774\n",
      "874\n",
      "988\n",
      "1062\n",
      "1063\n",
      "1191\n",
      "1261\n",
      "1262\n",
      "1413\n",
      "1484\n",
      "1487\n",
      "1490\n",
      "1633\n",
      "1634\n",
      "1637\n",
      "1699\n",
      "1700\n",
      "1701\n",
      "1815\n",
      "1816\n",
      "1876\n",
      "1878\n",
      "2003\n",
      "2051\n",
      "2054\n",
      "2183\n",
      "2235\n",
      "2377\n",
      "2428\n",
      "2549\n",
      "2552\n",
      "2589\n",
      "2590\n",
      "2682\n",
      "2728\n",
      "2729\n",
      "2730\n",
      "2804\n",
      "2832\n",
      "2834\n",
      "2836\n",
      "2912\n",
      "2913\n",
      "2935\n",
      "2938\n",
      "2939\n",
      "3028\n",
      "3064\n",
      "3067\n",
      "3069\n",
      "3170\n",
      "3200\n",
      "3258\n",
      "3298\n",
      "3301\n",
      "3303\n",
      "3418\n",
      "3444\n",
      "3524\n",
      "3525\n",
      "3541\n",
      "3598\n",
      "3623\n",
      "3625\n",
      "3628\n",
      "3704\n",
      "3748\n",
      "3749\n",
      "3860\n",
      "3862\n",
      "3873\n",
      "3876\n",
      "3877\n",
      "3878\n",
      "3889\n",
      "3890\n"
     ]
    }
   ],
   "source": [
    "pd.options.mode.chained_assignment = None\n",
    "# group_sv_by_gene(merged_df.head(5000))\n",
    "in_df_nogenes, out_df_nogenes = group_sv_by_gene_single_cohort_nogenes(cleaned_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_df_nogenes.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\cohort_clustering_nogenes\\GM24385_case_grouped_sv_out.csv\", index=False)\n",
    "out_df_nogenes.to_csv(rf\"Output\\output_hg002_clustering_04Jul2023\\cohort_clustering_nogenes\\GM24385_case_grouped_cluster_out.csv\", index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "breakdown of SV by type shared to AP for double checking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64,\n",
       " insertion               2601\n",
       " deletion                1219\n",
       " duplication               27\n",
       " duplication_inverted      24\n",
       " inversion_paired          13\n",
       " inversion                  3\n",
       " duplication_split          3\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_df_nogenes['Type'].value_counts(),out_df_nogenes['Type'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "compare solve vs enrichment clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['inversion_paired', 'inversion', 'insertion', 'deletion',\n",
       "       'duplication_inverted', 'duplication', 'duplication_split'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_sv_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "97\n",
      "98\n",
      "102\n",
      "303\n",
      "305\n",
      "405\n",
      "407\n",
      "568\n",
      "569\n",
      "641\n",
      "773\n",
      "873\n",
      "986\n",
      "1060\n",
      "1061\n",
      "1188\n",
      "1258\n",
      "1259\n",
      "1410\n",
      "1481\n",
      "1484\n",
      "1487\n",
      "1629\n",
      "1630\n",
      "1633\n",
      "1695\n",
      "1696\n",
      "1697\n",
      "1811\n",
      "1812\n",
      "1872\n",
      "1874\n",
      "1999\n",
      "2047\n",
      "2050\n",
      "2179\n",
      "2231\n",
      "2373\n",
      "2424\n",
      "2545\n",
      "2548\n",
      "2584\n",
      "2585\n",
      "2677\n",
      "2723\n",
      "2724\n",
      "2725\n",
      "2799\n",
      "2827\n",
      "2829\n",
      "2831\n",
      "2907\n",
      "2908\n",
      "2930\n",
      "2933\n",
      "2934\n",
      "3023\n",
      "3059\n",
      "3062\n",
      "3064\n",
      "3164\n",
      "3194\n",
      "3252\n",
      "3292\n",
      "3295\n",
      "3297\n",
      "3412\n",
      "3438\n",
      "3518\n",
      "3519\n",
      "3535\n",
      "3592\n",
      "3617\n",
      "3619\n",
      "3622\n",
      "3698\n",
      "3742\n",
      "3743\n",
      "3854\n",
      "3856\n",
      "3867\n",
      "3870\n",
      "3871\n",
      "3872\n",
      "3883\n",
      "3884\n"
     ]
    }
   ],
   "source": [
    "solve_in_df_nogenes, solve_out_df_nogenes = group_sv_by_gene_single_cohort_nogenes(solve_sv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(insertion               4337\n",
       " deletion                1751\n",
       " duplication               59\n",
       " duplication_inverted      53\n",
       " inversion_paired          21\n",
       " inversion                  6\n",
       " duplication_split          6\n",
       " Name: Type, dtype: int64,\n",
       " insertion               2597\n",
       " deletion                1217\n",
       " duplication               27\n",
       " duplication_inverted      24\n",
       " inversion_paired          13\n",
       " inversion                  3\n",
       " duplication_split          3\n",
       " Name: Type, dtype: int64)"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_in_df_nogenes['Type'].value_counts(),solve_out_df_nogenes['Type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "solve_cID = solve_in_df_nogenes['clusterId'].unique()\n",
    "# cohort_cID = solve_in_df['cluster_ID']\n",
    "\n",
    "discordant = []\n",
    "for i in solve_cID:\n",
    "    IDs = solve_in_df_nogenes.loc[solve_in_df_nogenes['clusterId'] == i]['cluster_ID']\n",
    "    if len(set(IDs)) > 1:\n",
    "        discordant.append(i)\n",
    "\n",
    "print(len(discordant))\n",
    "\n",
    "# mark rows with discordant cluster IDs\n",
    "solve_in_df_nogenes['discordant'] = solve_in_df_nogenes.apply(lambda x: True if x['clusterId'] in (discordant) else False, axis=1)\n",
    "solve_in_df_nogenes.loc[solve_in_df_nogenes['discordant'] == True].to_excel(rf\"Output\\output_hg002_clustering_04Jul2023\\solve_clustering\\GM24385_solve_discordant_cluster.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_cID = solve_in_df_nogenes['cluster_ID'].unique()\n",
    "# cohort_cID = solve_in_df['cluster_ID']\n",
    "\n",
    "discordant = []\n",
    "for i in solve_cID:\n",
    "    IDs = solve_in_df_nogenes.loc[solve_in_df_nogenes['cluster_ID'] == i]['clusterId']\n",
    "    if len(set(IDs)) > 1:\n",
    "        discordant.append(i)\n",
    "\n",
    "len(discordant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[981, 2728]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "discordant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusterId</th>\n",
       "      <th>SmapID</th>\n",
       "      <th>QryContigID</th>\n",
       "      <th>chr</th>\n",
       "      <th>RefcontigID2</th>\n",
       "      <th>QryStartPos</th>\n",
       "      <th>QryEndPos</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>RawConfidenceCenter</th>\n",
       "      <th>SVsize</th>\n",
       "      <th>SVfreq</th>\n",
       "      <th>orientation</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>has_gaps</th>\n",
       "      <th>OverlapGenes</th>\n",
       "      <th>clustered</th>\n",
       "      <th>cluster_ID</th>\n",
       "      <th>discordant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1365</th>\n",
       "      <td>2635</td>\n",
       "      <td>4889</td>\n",
       "      <td>192</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>364190.4</td>\n",
       "      <td>374066.1</td>\n",
       "      <td>189831786.5</td>\n",
       "      <td>189838877.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>70.88</td>\n",
       "      <td>2785.2</td>\n",
       "      <td>0.391</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>981</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1366</th>\n",
       "      <td>2636</td>\n",
       "      <td>4897</td>\n",
       "      <td>2842</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>395375.2</td>\n",
       "      <td>403306.8</td>\n",
       "      <td>189845937.0</td>\n",
       "      <td>189849826.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>50.05</td>\n",
       "      <td>4042.6</td>\n",
       "      <td>0.268</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>981</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>2636</td>\n",
       "      <td>4698</td>\n",
       "      <td>191</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>395760.8</td>\n",
       "      <td>403698.2</td>\n",
       "      <td>189845937.0</td>\n",
       "      <td>189849826.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>872.34</td>\n",
       "      <td>4048.3</td>\n",
       "      <td>0.320</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>981</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     clusterId SmapID QryContigID  chr RefcontigID2 QryStartPos QryEndPos  \\\n",
       "1365      2635   4889         192    4            4    364190.4  374066.1   \n",
       "1366      2636   4897        2842    4            4    395375.2  403306.8   \n",
       "1367      2636   4698         191    4            4    395760.8  403698.2   \n",
       "\n",
       "      RefStartPos    RefEndPos Confidence  ... RawConfidenceCenter  SVsize  \\\n",
       "1365  189831786.5  189838877.0       0.99  ...               70.88  2785.2   \n",
       "1366  189845937.0  189849826.0       0.99  ...               50.05  4042.6   \n",
       "1367  189845937.0  189849826.0       0.99  ...              872.34  4048.3   \n",
       "\n",
       "     SVfreq orientation Sample_ID has_gaps OverlapGenes clustered cluster_ID  \\\n",
       "1365  0.391         NA    GM24385    False            -      True        981   \n",
       "1366  0.268         NA    GM24385    False            -      True        981   \n",
       "1367  0.320         NA    GM24385    False            -      True        981   \n",
       "\n",
       "     discordant  \n",
       "1365      False  \n",
       "1366       True  \n",
       "1367       True  \n",
       "\n",
       "[3 rows x 34 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_in_df_nogenes.loc[solve_in_df_nogenes['cluster_ID'] == 981]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusterId</th>\n",
       "      <th>SmapID</th>\n",
       "      <th>QryContigID</th>\n",
       "      <th>chr</th>\n",
       "      <th>RefcontigID2</th>\n",
       "      <th>QryStartPos</th>\n",
       "      <th>QryEndPos</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>RawConfidenceCenter</th>\n",
       "      <th>SVsize</th>\n",
       "      <th>SVfreq</th>\n",
       "      <th>orientation</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>has_gaps</th>\n",
       "      <th>OverlapGenes</th>\n",
       "      <th>clustered</th>\n",
       "      <th>cluster_ID</th>\n",
       "      <th>discordant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4397</th>\n",
       "      <td>6712</td>\n",
       "      <td>12581</td>\n",
       "      <td>1371</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>361934.6</td>\n",
       "      <td>378545.7</td>\n",
       "      <td>19009467.0</td>\n",
       "      <td>19022996.5</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>71.59</td>\n",
       "      <td>3081.6</td>\n",
       "      <td>0.062</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>2733</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4398</th>\n",
       "      <td>6712</td>\n",
       "      <td>12590</td>\n",
       "      <td>1591</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>217718.6</td>\n",
       "      <td>224135.0</td>\n",
       "      <td>19008159.0</td>\n",
       "      <td>19013073.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>75.92</td>\n",
       "      <td>1502.4</td>\n",
       "      <td>0.023</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>2732</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4399</th>\n",
       "      <td>6712</td>\n",
       "      <td>12585</td>\n",
       "      <td>1372</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>361934.6</td>\n",
       "      <td>376943.7</td>\n",
       "      <td>19009467.0</td>\n",
       "      <td>19022996.5</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>37.33</td>\n",
       "      <td>1479.6</td>\n",
       "      <td>0.050</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>2732</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4400</th>\n",
       "      <td>6712</td>\n",
       "      <td>12566</td>\n",
       "      <td>1602</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>524392.1</td>\n",
       "      <td>539538.2</td>\n",
       "      <td>19009467.0</td>\n",
       "      <td>19022996.5</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>44.82</td>\n",
       "      <td>1616.6</td>\n",
       "      <td>0.017</td>\n",
       "      <td>NA</td>\n",
       "      <td>GM24385</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>2732</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     clusterId SmapID QryContigID  chr RefcontigID2 QryStartPos QryEndPos  \\\n",
       "4397      6712  12581        1371   14           14    361934.6  378545.7   \n",
       "4398      6712  12590        1591   14           14    217718.6  224135.0   \n",
       "4399      6712  12585        1372   14           14    361934.6  376943.7   \n",
       "4400      6712  12566        1602   14           14    524392.1  539538.2   \n",
       "\n",
       "      RefStartPos   RefEndPos Confidence  ... RawConfidenceCenter  SVsize  \\\n",
       "4397   19009467.0  19022996.5       0.99  ...               71.59  3081.6   \n",
       "4398   19008159.0  19013073.0       0.99  ...               75.92  1502.4   \n",
       "4399   19009467.0  19022996.5       0.99  ...               37.33  1479.6   \n",
       "4400   19009467.0  19022996.5       0.99  ...               44.82  1616.6   \n",
       "\n",
       "     SVfreq orientation Sample_ID has_gaps OverlapGenes clustered cluster_ID  \\\n",
       "4397  0.062         NA    GM24385    False            -     False       2733   \n",
       "4398  0.023         NA    GM24385    False            -     False       2732   \n",
       "4399  0.050         NA    GM24385    False            -     False       2732   \n",
       "4400  0.017         NA    GM24385    False            -     False       2732   \n",
       "\n",
       "     discordant  \n",
       "4397       True  \n",
       "4398       True  \n",
       "4399       True  \n",
       "4400       True  \n",
       "\n",
       "[4 rows x 34 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_in_df_nogenes.loc[solve_in_df_nogenes['clusterId'] == '6712']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "enrichment",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
