{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Cohort SV Enrichment Project**\n",
    "\n",
    "**Author:** Shuk\n",
    "\n",
    "**Background**\n",
    "Following up on `Augusta_Layman` SV enrichment project, an analysis workflow suggestion was to calculate percent presence of a controlb SV within itself. This function will be tested and compared against an .smap file's `Present_in_%_of_BNG_control_samples`and `Present_in_%_of_BNG_control_samples_with_the_same_enzyme` columns.\n",
    "\n",
    "**Objective**\n",
    "- Implement a function that takes an SV as input, finds alignments/non-alignments and calculates percent presence in % of BNG control samples.\n",
    "\n",
    "**Concept**\n",
    "- given an SV, find similar (poswindow, reciprocalsize) SVs like it in controldb. Look into the unique Sample IDs of these similar SVs, then divide that by total number ofd unique sample IDs in controldb"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TO-DO**\n",
    "1)  Use import os to improve internal file calling, no need to input root path.\n",
    "2)  Case SV selection criteria: shared by >2 SVs and overlaps morbid gene. Match case to control SVs.\n",
    "3)  Implement functions to gradually add new SVs into existing groups/create new SVs as new .smaps are added\n",
    "4)  Add DGV bed file filter\n",
    "5)  Add inversion flattening feature\n",
    "6)  Make case and control SV filter summary for QC"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RUN THIS FIRST: Load Python packages, BED files, and helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load packages\n",
    "from itertools import groupby\n",
    "from matplotlib_venn import venn2, venn2_circles\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.stats import chi2\n",
    "from scipy.stats import chisquare\n",
    "from scipy.stats import ttest_ind\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import f_oneway\n",
    "# from scipy.stats.contingency import odds_ratio\n",
    "\n",
    "import copy\n",
    "import re \n",
    "import os\n",
    "import pathlib\n",
    "working_dir = pathlib.Path().absolute()\n",
    "os.chdir(working_dir)\n",
    "\n",
    "\n",
    "%matplotlib inline\n",
    "pd.options.mode.chained_assignment = None  # default='warn'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify columns needed for output list\n",
    "out_headers = [\"OverlapGenes\", \"RefcontigID1\" , \"RefcontigID2\", \n",
    "                \"RefStartPos\", \"RefEndPos\", \"Type\", \"Zygosity\", \n",
    "                \"Present_in_%_of_BNG_control_samples\", \"Present_in_%_of_BNG_control_samples_with_the_same_enzyme\", \n",
    "                \"SVsize\", \"SVfreq\", \"Self_molecule_count\"]        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to get sample ID\n",
    "def get_sample_name(sample_list, filename):\n",
    "    # Sample ID \n",
    "    sample_ID = filename.replace(\"_smap_table.txt\",\"\")\n",
    "    \n",
    "    # check is sample ID is in sample list, return if true\n",
    "    if sample_list[\"Sample Name\"].str.contains(sample_ID).any():\n",
    "        return sample_ID\n",
    "\n",
    "# TEST\n",
    "# print(get_sample_name(ntd_110_list, '17420_smap_table.txt'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to filter for SVs equal or below a certain threshold (inadvertently excludes inversion partials)\n",
    "def filter_rare_svs(df, percent_cutoff):\n",
    "    df = df.astype({\"Present_in_%_of_BNG_control_samples\":'float64', \"Present_in_%_of_BNG_control_samples_with_the_same_enzyme\":'float64'})\n",
    "\n",
    "    return df.loc[((df[\"Present_in_%_of_BNG_control_samples\"] <= percent_cutoff) & (df[\"Present_in_%_of_BNG_control_samples_with_the_same_enzyme\"] <= percent_cutoff))\n",
    "            & ((df[\"Present_in_%_of_BNG_control_samples\"] != -1) & (df[\"Present_in_%_of_BNG_control_samples_with_the_same_enzyme\"] != -1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def exclude_types_svs(df):\n",
    "    excluded_types = [\"deletion_nbase\", \"insertion_nbase\", \"gain_masked\", \"loss_masked\", \"inversion_partial\", \"trans_interchr_common\", \"trans_intrachr_common\", 'translocation_interchr', 'trans_intrachr_segdupe', 'trans_interchr_segdupe', 'translocation_intrachr']\n",
    "    return df.loc[~df[\"Type\"].isin(excluded_types)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output file for QC\n",
    "# File has SVs without trans removed, n_gaps, and overlapping genes labeled\n",
    "def out_file(df, name):\n",
    "    df.to_csv(rf\"Output\\percent_presence_28Feb2023\\{name}.csv\", index=False)\n",
    "\n",
    "def in_file(name):\n",
    "    df = pd.read_csv(rf\"Output\\percent_presence_28Feb2023\\{name}.csv\", dtype=object, index_col=False)\n",
    "    return df\n",
    "\n",
    "# out_file(ctrldb_df, \"ctrldb_df_2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ['deletion', 'insertion', 'insertion_nbase',\n",
    "#        'trans_interchr_common', 'trans_intrachr_common',\n",
    "#        'translocation_interchr', 'duplication', 'inversion',\n",
    "#        'trans_intrachr_segdupe', 'duplication_inverted', 'deletion_nbase',\n",
    "#        'inversion_paired', 'trans_interchr_segdupe', 'duplication_split',\n",
    "#        'translocation_intrachr']\n",
    "\n",
    "# Helper function to filter by confidence intervals\n",
    "\n",
    "# 'trans', 'duplication_inverted', 'duplication','duplication_split', 'inv', 'ins', 'del'\n",
    "def filter_sig_svs(df):\n",
    "    dict_conf = {\n",
    "        \"ins\": 0,\n",
    "        \"del\": 0,\n",
    "        \"dup\": -1,\n",
    "        \"inv\": float(0.7),\n",
    "        \"trans\": float(0.05) # intrachr and interchr\n",
    "    }\n",
    "\n",
    "    def get_sv_conf_thres(sv):\n",
    "        sv = str(sv)\n",
    "        if \"ins\" in sv:\n",
    "            return dict_conf[\"ins\"]\n",
    "        elif \"del\" in sv:\n",
    "            return dict_conf[\"del\"]\n",
    "        elif \"duplication\" in sv:\n",
    "            return dict_conf[\"dup\"]\n",
    "        elif \"inv\" in sv:\n",
    "            return dict_conf[\"inv\"]\n",
    "        elif \"trans\" in sv:\n",
    "            return dict_conf[\"trans\"]\n",
    "        else:\n",
    "            print(f\"unknown sv type: {sv}\")\n",
    "\n",
    "    df = df.astype({'Confidence':'float64'})\n",
    "    pass_conf = df[[\"Confidence\", \"Type\"]].apply(lambda x: True if x[0] >= float(get_sv_conf_thres(x[1])) else False, axis=1)\n",
    "\n",
    "    return df.loc[pass_conf == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper function to flatten inversion pairs into 1 inversion\n",
    "def flatten_df_inv(df):\n",
    "    # count inversion pairs and singles\n",
    "    n_inv, n_pair, n_sing = 0,0,0\n",
    "\n",
    "    # subset inversions and inversion_partials\n",
    "    inv = ['inversion', 'inversion_partial']\n",
    "    inv_df = copy.deepcopy(df.loc[df['Type'].isin(inv)])\n",
    "\n",
    "    n_inv = len(inv_df)\n",
    "    print(\"inversions & inversion_partials:\", n_inv)\n",
    "\n",
    "    # initialize empty dataframe\n",
    "    out_df = pd.DataFrame()\n",
    "\n",
    "    # proceed if there are inversions, return original df if there are none\n",
    "    if len(inv_df) == 0:\n",
    "        return df\n",
    "    \n",
    "    # remove inversions from main df\n",
    "    df = df.loc[~df['Type'].isin(inv)]\n",
    "\n",
    "    # initialize column of T/F, where SVs == T have been analyzed for flattening, and F haven't\n",
    "    inv_df.loc[:, 'flattened'] =  False\n",
    "\n",
    "    # loops through inversions and flattens inversions where LinkID matches SmapID\n",
    "    while False in pd.unique(inv_df['flattened']):\n",
    "        \n",
    "        df_copy = copy.deepcopy(inv_df.loc[inv_df['flattened'] == False])\n",
    "        # print(len(df_copy))\n",
    "\n",
    "        # get first row item\n",
    "        qry_ID = df_copy['Sample_ID'].iat[0]\n",
    "        qry_chr = df_copy['chr'].iat[0]\n",
    "        qry_smapID = df_copy['SmapID'].iat[0]\n",
    "        qry_linkID = df_copy['LinkID'].iat[0]\n",
    "\n",
    "        # find pair for the row if exists\n",
    "        df_copy = df_copy.loc[(df_copy['Sample_ID'] == qry_ID)\n",
    "                            & (df_copy['chr'] == qry_chr)\n",
    "                            & ((df_copy['SmapID'] == qry_linkID)\n",
    "                            | (df_copy['SmapID'] == qry_smapID))]\n",
    "\n",
    "        # update original inv_df column as flattened\n",
    "        inv_df.loc[(inv_df['Sample_ID'] == qry_ID) \n",
    "                    & (inv_df['chr'] == qry_chr) \n",
    "                    & ((inv_df['SmapID'] == qry_linkID) \n",
    "                    | (inv_df['SmapID'] == qry_smapID)), ['flattened']] = True\n",
    "\n",
    "        # flatten if there is a pair, report only row if not\n",
    "        if len(df_copy) > 1:\n",
    "            \n",
    "            # increment counter\n",
    "            n_pair += 1\n",
    "\n",
    "            # choose inversion in the pair, combine both\n",
    "            if \"inversion\" in pd.unique(df_copy['Type']):\n",
    "                row = df_copy.loc[(df_copy['RefEndPos'] != -1.0) & (df_copy['Type'] == 'inversion')].reset_index() # choose inversions\n",
    "            else:\n",
    "                row = df_copy.iloc[[0]].reset_index()\n",
    "            \n",
    "            # merge pairs, choosing the lowest and largest\n",
    "            ref = np.concatenate((df_copy['RefStartPos'].to_numpy(), df_copy['RefEndPos'].to_numpy()), axis=None)\n",
    "            ref = np.delete(ref, np.where(ref == -1))\n",
    "            # print(ref)\n",
    "\n",
    "            row.at[0, 'RefStartPos'] = np.amin(ref, axis=None)\n",
    "            row.at[0, 'RefEndPos'] = np.amax(ref, axis=None)\n",
    "\n",
    "        else:\n",
    "            # increment counter\n",
    "            n_sing += 1\n",
    "            \n",
    "            row = df_copy.iloc[[0]].reset_index()\n",
    "            \n",
    "\n",
    "        # append to out_df        \n",
    "        # print(row)\n",
    "        out_df = pd.concat([out_df, row], axis=0, ignore_index=True)\n",
    "\n",
    "    # print flattened inversions for QC\n",
    "    out_file(out_df, 'flattened_inv_out')\n",
    "\n",
    "    # append out_df to df\n",
    "\n",
    "    # return concatenated flattened inversions with original input df\n",
    "    return pd.concat([out_df[df.columns], df], axis=0, ignore_index=True), n_inv, n_pair, n_sing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standardized headers for both bed files\n",
    "# headers: chrom\tchromStart\tchromEnd\tGene\tIndex\tstrand\tchromStart2\tchromEnd\tRGB\n",
    "bed_headers = [\"chr\",\"RefStartPos\",\"RefEndPos\",\"Gene\",\"Index\",\"Strand\",\"RefStartPos2\",\"RefEndPos2\",\"RGB\"]\n",
    "\n",
    "# Load hg38.knownCanonical.bed\n",
    "bed_df = pd.read_csv(r\"Input\\ref_genes\\hg38.knownCanonical.mapped_KH.bed\", dtype=object, index_col=False, sep=\"\\t\")\n",
    "bed_df = bed_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "\n",
    "# Load hg38_gaps.bed\n",
    "gaps_df = pd.read_csv(r\"Input\\ref_genes\\hg38_gaps.bed\", dtype=object,  index_col=False, sep=\"\\t\", names=bed_headers)\n",
    "gaps_df = gaps_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "\n",
    "# Load hg38 pseudogene list.bed\n",
    "pseudo_df = pd.read_csv(r\"Input\\ref_genes\\hg38_EncodeGencodePseudoGeneV41.bed\",dtype=object,  index_col=False, sep=\"\\t\")\n",
    "pseudo_df = pseudo_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "\n",
    "# set redstartstop as float for downstream filtering functions\n",
    "for df in [bed_df, gaps_df, pseudo_df]:\n",
    "    df = df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int'})\n",
    "\n",
    "# display(bed_df.head(3))\n",
    "# display(gaps_df.head(3))\n",
    "# bed_df.shape, gaps_df.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to remove SVs overlapping gaps bed file\n",
    "def filter_genes(bed_df, chr, start, end):\n",
    "    \n",
    "    match_chr = bed_df.loc[(bed_df[\"chr\"] == chr)]\n",
    "\n",
    "    # display(match_chr)\n",
    "\n",
    "    # transform matched chr rows into numpy (faster processing than directly panda slicing)\n",
    "    bed_genes = np.array(match_chr[\"Gene\"])\n",
    "    bed_start = np.array(match_chr[\"RefStartPos\"])\n",
    "    bed_end = np.array(match_chr[\"RefEndPos\"])\n",
    "\n",
    "    overlap_idx = np.where((((start <= bed_start) & (bed_start <= end)) \n",
    "                            | ((bed_start <= start) & (start <= bed_end))))\n",
    "    \n",
    "    overlap_genes = bed_genes[overlap_idx]\n",
    "    \n",
    "    if len(overlap_genes) > 0:\n",
    "        return \";\".join(list(np.unique(overlap_genes)))\n",
    "    else:\n",
    "        return \"-\"\n",
    "\n",
    "# filter by chr, RefStartPos, and RefEndPos\n",
    "# filter_nogaps(9, 41132473.0, 41293385.0) # Test\n",
    "\n",
    "# Helper function to remove SVs overlapping gaps bed file\n",
    "def filter_nogaps(type, chr, start, end):\n",
    "\n",
    "    # n_gaps check only on insertion and deletions\n",
    "    if ((\"ins\" in type) | (\"del\" in type)): \n",
    "        match_chr = gaps_df.loc[(gaps_df[\"chr\"] == chr)]\n",
    "        # check each SV related to gaps[chr\"] == chr for overlaps, False if none\n",
    "\n",
    "        # transform matched chr rows into numpy (faster processing than directly panda slicing)\n",
    "        # bed_genes = np.array(match_chr[\"Gene\"])\n",
    "        bed_start = np.array(match_chr[\"RefStartPos\"])\n",
    "        bed_end = np.array(match_chr[\"RefEndPos\"])\n",
    "\n",
    "        for bedStart, bedEnd in zip(bed_start, bed_end):\n",
    "            if (((start <= bedStart) & (bedStart <= end)) \n",
    "            | ((bedStart <= start) & (start <= bedEnd))):\n",
    "                return True\n",
    "\n",
    "    return False\n",
    "\n",
    "# filter by chr, RefStartPos, and RefEndPos\n",
    "# filter_nogaps(\"ins\", 9, 41132473.0, 41293385.0) # Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_sorted_unique_genes(gene_col):\n",
    "    return gene_col.str.split(';').map(lambda x: ','.join(list(set(sorted(x)))))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load de novo Controlsdb (Control)\n",
    "\n",
    "This section is taken from `layman_enrichment_analysis_singlebkpt_nonmasked_500kb`"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load ctrldb for de novo assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample', 'Type', 'algorithm', 'chr', 'bkpt', 'chrB', 'bkptB',\n",
       "       'RefStartPos', 'RefEndPos', 'Size', 'Zygosity', 'Confidence', 'smapID',\n",
       "       'linkSmapId', 'has_gaps'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl_df = in_file('ctrl_nogaps_out').astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'Size':'float64'})\n",
    "ctrl_df = ctrl_df.loc[ctrl_df['Sample'].str.contains(\"dle1\")]\n",
    "ctrl_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2749893"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ctrl_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['duplication_inverted', 'duplication', 'duplication_split', 'inv',\n",
       "       'ins', 'del'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "179"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ctrl_df[ctrl_df['Sample'].str.contains('dle')]['Sample'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ctrl_df = in_file('control_sig_out').astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'Size':'float64'})\n",
    "# ctrl_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Load de novo controlsdb and standardize headers\n",
    "# ctrldb_df = pd.read_csv(r\"Input\\controlsdb\\ctrl_sv_db_anonymize_hg38.txt\", index_col=False, sep=\"\\t\")\n",
    "# ctrldb_headers = [\"Sample\", \"Type\", \"algorithm\", \"chr\", \"bkpt\", \"chrB\", \"bkptB\", \"RefStartPos\", \"RefEndPos\", \"Size\", \"Zygosity\", \"Confidence\", \"smapID\", \"linkSmapId\"]\n",
    "# ctrldb_df.columns = ctrldb_headers\n",
    "# ctrldb_df = ctrldb_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'Size':'float64'})\n",
    "# ctrldb_df = ctrldb_df.loc[~ctrldb_df['Type'].str.contains(\"trans\")]\n",
    "# ctrldb_df[\"has_gaps\"] = ctrldb_df.apply(lambda x: filter_nogaps(x[\"Type\"], x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "# ctrldb_df = ctrldb_df.loc[ctrldb_df[\"has_gaps\"] == False]\n",
    "# ctrldb_df = ctrldb_df.loc[ctrl_df['Sample'].str.contains(\"dle1\")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file(ctrldb_df, \"ctrl_nogaps_out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file(pd.DataFrame(ctrldb_df['Sample'].unique()), \"all_ctrldb_samples\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "filter for dle1 samples only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrldb_df = ctrldb_df.loc[ctrldb_df['Sample'].str.contains(\"dle1\")]\n",
    "\n",
    "# # filter controlsdb by score filter\n",
    "# sig_df = filter_sig_svs(dle1_df)\n",
    "# print(f\"{len(sig_df)} SVs are significant, with confidence scores above Access recommended threshold\")\n",
    "\n",
    "# out_file(sig_df, \"control_sig_out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample', 'Type', 'algorithm', 'chr', 'bkpt', 'chrB', 'bkptB',\n",
       "       'RefStartPos', 'RefEndPos', 'Size', 'Zygosity', 'Confidence', 'smapID',\n",
       "       'linkSmapId'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ctrldb_df = in_file('control_sig_out').astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'Size':'float64'})\n",
    "# ctrldb_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample', 'Type', 'algorithm', 'chr', 'bkpt', 'chrB', 'bkptB',\n",
       "       'RefStartPos', 'RefEndPos', 'Size', 'Zygosity', 'Confidence', 'smapID',\n",
       "       'linkSmapId'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df.columns"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get % Presence in Controldb"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Translocation coordinates are marked based on `bkpt` and `bkptB`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['trans'], dtype=object)"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df[~ctrldb_df['bkpt'].isna()]['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['trans', 'duplication_inverted', 'duplication',\n",
       "       'duplication_split', 'inv', 'ins', 'del'], dtype=object)"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['trans'], dtype=object)"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df[ctrldb_df['Type'].str.contains('trans')]['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7359682.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7950679.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7408</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7359682.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7950679.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7409</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7913502.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7398088.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7417</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7807846.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7513171.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7421</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7913502.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7398088.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7422</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample   Type            algorithm  chr       bkpt  chrB      bkptB  \\\n",
       "0  1_dle1_sop  trans  assembly_comparison    8  7359682.0   8.0  7950679.0   \n",
       "1  1_dle1_sop  trans  assembly_comparison    8  7359682.0   8.0  7950679.0   \n",
       "2  1_dle1_sop  trans  assembly_comparison    8  7913502.0   8.0  7398088.0   \n",
       "3  1_dle1_sop  trans  assembly_comparison    8  7807846.0   8.0  7513171.0   \n",
       "4  1_dle1_sop  trans  assembly_comparison    8  7913502.0   8.0  7398088.0   \n",
       "\n",
       "   RefStartPos  RefEndPos  Size      Zygosity  Confidence  smapID  linkSmapId  \n",
       "0          NaN        NaN   NaN  heterozygous        0.01    7408         NaN  \n",
       "1          NaN        NaN   NaN  heterozygous        0.01    7409         NaN  \n",
       "2          NaN        NaN   NaN  heterozygous        0.00    7417         NaN  \n",
       "3          NaN        NaN   NaN    homozygous        0.01    7421         NaN  \n",
       "4          NaN        NaN   NaN  heterozygous        0.00    7422         NaN  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "179\n"
     ]
    }
   ],
   "source": [
    "# Helper function to calculate % presence in controldb\n",
    "ref_ctrl_df = in_file('ctrl_nogaps_out').astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'Size':'float64'})\n",
    "# ref_ctrl_df = in_file('control_sig_out').astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'Size':'float64'})\n",
    "# ref_ctrl_df =pd.read_csv(r\"Input\\controlsdb\\ctrl_sv_db_anonymize_hg38.txt\", index_col=False, sep=\"\\t\")\n",
    "# ctrldb_headers = [\"Sample\", \"Type\", \"algorithm\", \"chr\", \"bkpt\", \"chrB\", \"bkptB\", \"RefStartPos\", \"RefEndPos\", \"Size\", \"Zygosity\", \"Confidence\", \"smapID\", \"linkSmapId\"]\n",
    "# ref_ctrl_df.columns = ctrldb_headers\n",
    "# ctrldb_df = ctrldb_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int','Size':'float64'})\n",
    "\n",
    "ref_ctrl_df = ref_ctrl_df.loc[ref_ctrl_df['Sample'].str.contains(\"dle1\")]\n",
    "ref_ctrl_df = ref_ctrl_df.loc[~ref_ctrl_df['Type'].str.contains(\"trans\")]\n",
    "ref_ctrl_df.columns\n",
    "\n",
    "ctrl_db_n = len(ref_ctrl_df.loc[ref_ctrl_df['Sample'].str.contains(\"dle1\")]['Sample'].unique())\n",
    "print(ctrl_db_n)\n",
    "\n",
    "def calculate_percent_presence_controldb(Type, zygo, chrA, chrB, start, end, size):\n",
    "\n",
    "    def get_sv_type(sv):\n",
    "        sv = str(sv)\n",
    "        if \"ins\" in sv:\n",
    "            return \"ins\"\n",
    "        elif \"del\" in sv:\n",
    "            return \"del\"\n",
    "        elif \"duplication\" in sv:\n",
    "            return sv\n",
    "        elif \"inv\" in sv:\n",
    "            return \"inv\"\n",
    "        elif \"trans\" in sv:\n",
    "            return \"trans\"\n",
    "        else:\n",
    "            print(f\"unknown sv type: {sv}\")\n",
    "\n",
    "    # transform matched chr rows into numpy (faster processing than directly panda slicing)\n",
    "    # bed_genes = np.array(match_chr[\"Gene\"])\n",
    "\n",
    "    Type = get_sv_type(Type)\n",
    "\n",
    "    ins_del_position_overlap=10000\n",
    "    ins_del_size_percent_similarity=50\n",
    "    inversion_position_overlap=50000\n",
    "    translocation_position_overlap=50000\n",
    "    duplication_position_overlap=10000\n",
    "    duplication_size_percent_similarity=50\n",
    "\n",
    "    if (Type == 'ins') | (Type == 'del'):\n",
    "        posWindow, reciprocalSize = ins_del_position_overlap, ins_del_size_percent_similarity\n",
    "    elif 'duplication' in Type:\n",
    "        posWindow, reciprocalSize = duplication_position_overlap, duplication_size_percent_similarity\n",
    "    elif Type == 'inv':\n",
    "        posWindow, reciprocalSize = inversion_position_overlap, 0\n",
    "    elif Type == 'trans':  \n",
    "        posWindow, reciprocalSize = translocation_position_overlap, 0\n",
    "\n",
    "    bed_df = ref_ctrl_df.copy()\n",
    "\n",
    "    # if interchr_translocation\n",
    "    if Type == 'trans':\n",
    "        print(Type)\n",
    "        if chrA == chrB:\n",
    "        \n",
    "            match_chr = bed_df.loc[(bed_df[\"chr\"] == chrA) & (bed_df['Type'] == Type)]\n",
    "\n",
    "            bed_sample = np.array(match_chr['Sample'])\n",
    "            bed_start = np.array(match_chr[\"bkpt\"])\n",
    "            bed_end = np.array(match_chr[\"bkptB\"])\n",
    "\n",
    "            overlap_idx = np.where((((start - posWindow <= bed_start) & (bed_start <= end + posWindow)) \\\n",
    "                                    | ((bed_start <= start - posWindow) & (start - posWindow <= bed_end))))\n",
    "\n",
    "            percent_presence = 100 * len(list(match_chr.iloc[overlap_idx]['Sample'].unique()))/ctrl_db_n\n",
    "            \n",
    "        else:\n",
    "            match_chrA = bed_df.loc[(bed_df[\"chr\"] == chrA) & (bed_df['Type'] == Type)]\n",
    "            match_chrB = bed_df.loc[(bed_df[\"chr\"] == chrB) & (bed_df['Type'] == Type)]\n",
    "\n",
    "            bed_start = np.array(match_chrA[\"bkpt\"])\n",
    "            bed_end = np.array(match_chrB[\"bkptB\"])\n",
    "\n",
    "            overlap_idxA = np.where((((start - posWindow <= bed_start) & (bed_start <= end + posWindow)) \\\n",
    "                                    | ((bed_start <= start - posWindow) & (start - posWindow <= bed_end))))\n",
    "            \n",
    "            overlap_idxB = np.where((((start - posWindow <= bed_start) & (bed_start <= end + posWindow)) \\\n",
    "                                    | ((bed_start <= start - posWindow) & (start - posWindow <= bed_end))))\n",
    "            \n",
    "            samples = pd.concat([match_chrA.iloc[overlap_idxA]['Sample'], match_chrB.iloc[overlap_idxB]['Sample']], axis=0, ignore_index=True)\n",
    "            n = list(samples.unique())\n",
    "            display(n)\n",
    "            percent_presence = 100 * len(n)/ctrl_db_n\n",
    "            \n",
    "    elif Type == 'inv':\n",
    "        print(Type)\n",
    "        \n",
    "        match_chr = bed_df.loc[(bed_df[\"chr\"] == chrA) & (bed_df['Type'] == Type)]\n",
    "\n",
    "        bed_sample = np.array(match_chr['Sample'])\n",
    "        bed_start = np.array(match_chr[\"RefStartPos\"])\n",
    "        bed_end = np.array(match_chr[\"RefEndPos\"])\n",
    "        bed_size = np.array(match_chr[\"Size\"])\n",
    "\n",
    "        overlap_idx = np.where((((start - posWindow <= bed_start) & (bed_start <= end + posWindow)) \\\n",
    "                                | ((bed_start <= start - posWindow) & (start - posWindow <= bed_end))))\n",
    "        \n",
    "        percent_presence = 100 * len(np.unique(bed_sample[overlap_idx]))/ctrl_db_n\n",
    "    # if indeldup\n",
    "    else:\n",
    "        print(Type)\n",
    "        \n",
    "        match_chr = bed_df.loc[(bed_df[\"chr\"] == chrA) & (bed_df['Type'] == Type)]\n",
    "\n",
    "        bed_sample = np.array(match_chr['Sample'])\n",
    "        bed_start = np.array(match_chr[\"RefStartPos\"])\n",
    "        bed_end = np.array(match_chr[\"RefEndPos\"])\n",
    "        bed_size = np.array(match_chr[\"Size\"])\n",
    "\n",
    "        overlap_idx = np.where((((start - posWindow <= bed_start) & (bed_start <= end + posWindow)) \\\n",
    "                                | ((bed_start <= start - posWindow) & (start - posWindow <= bed_end)))\n",
    "                                & (((size/bed_size)*100 >= reciprocalSize) \\\n",
    "                                    & ((bed_size/size)*100 >= reciprocalSize)))\n",
    "        \n",
    "        # display(match_chr.iloc[overlap_idx])\n",
    "        percent_presence = 100 * len(np.unique(bed_sample[overlap_idx]))/ctrl_db_n\n",
    "\n",
    "    # display(list(match_chr.iloc[overlap_idx]['Sample'].unique()))\n",
    "    return percent_presence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>2</td>\n",
       "      <td>134647984.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>140754758.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.05</td>\n",
       "      <td>2126</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>12938769.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13263012.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.32</td>\n",
       "      <td>105</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>12938769.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13263012.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.3</td>\n",
       "      <td>158</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>12938769.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13263012.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.07</td>\n",
       "      <td>184</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>13005792.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13223159.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.33</td>\n",
       "      <td>188</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample   Type            algorithm  chr         bkpt chrB        bkptB  \\\n",
       "0  1_dle1_sop  trans  assembly_comparison    2  134647984.0  2.0  140754758.0   \n",
       "1  1_dle1_sop  trans  assembly_comparison    1   12938769.0  1.0   13263012.0   \n",
       "2  1_dle1_sop  trans  assembly_comparison    1   12938769.0  1.0   13263012.0   \n",
       "3  1_dle1_sop  trans  assembly_comparison    1   12938769.0  1.0   13263012.0   \n",
       "4  1_dle1_sop  trans  assembly_comparison    1   13005792.0  1.0   13223159.0   \n",
       "\n",
       "   RefStartPos  RefEndPos  Size      Zygosity Confidence smapID linkSmapId  \n",
       "0          NaN        NaN   NaN  heterozygous       0.05   2126        NaN  \n",
       "1          NaN        NaN   NaN  heterozygous       0.32    105        NaN  \n",
       "2          NaN        NaN   NaN  heterozygous        0.3    158        NaN  \n",
       "3          NaN        NaN   NaN  heterozygous       0.07    184        NaN  \n",
       "4          NaN        NaN   NaN  heterozygous       0.33    188        NaN  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_ctrl_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load case cleaned file for testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_df = in_file('case_compiled_out')\n",
    "case_df.rename(columns={\"RefcontigID2\":\"chrB\"}, inplace=True)\n",
    "case_df.rename(columns={\"RefcontigID1\":\"chr\"}, inplace=True)\n",
    "case_df = case_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int', 'chrB':'int', 'num_overlap_DGV_calls':'int', 'SVsize':'float64', 'Present_in_%_of_BNG_control_samples_with_the_same_enzyme':'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_df[case_df['SVsize'] == 0]['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for type in case_df['Type'].unique():\n",
    "    if 'trans' in type:\n",
    "        continue\n",
    "    # df = case_df.loc[(case_df['Type'] == type) & (case_df['Present_in_%_of_BNG_control_samples_with_the_same_enzyme'] <= 1.0)].head()\n",
    "    df = case_df.loc[(case_df['Type'] == type)].head()\n",
    "    df['%_controldb'] = df[['Type', 'Zygosity','chr', 'chrB', 'RefStartPos', 'RefEndPos', 'SVsize']].apply(lambda x: calculate_percent_presence_controldb(x[0], x[1], x[2], x[3], x[4], x[5], x[6]), axis=1)\n",
    "    print(type), display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rare_df = case_df.loc[(case_df['Present_in_%_of_BNG_control_samples_with_the_same_enzyme'] <= 1.0)].copy()\n",
    "rare_df['%_controldb'] = rare_df[['Type', 'Zygosity','chr', 'chrB', 'RefStartPos', 'RefEndPos', 'SVsize']].apply(lambda x: calculate_percent_presence_controldb(x[0], x[1], x[2], x[3], x[4], x[5], x[6]), axis=1)\n",
    "out_file(case_df, 'test_percent_controldb_out')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filtering Controldb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load de novo controlsdb and standardize headers\n",
    "ctrldb_df = pd.read_csv(r\"Input\\controlsdb\\ctrl_sv_db_anonymize_hg38.txt\", index_col=False, sep=\"\\t\")\n",
    "ctrldb_headers = [\"Sample\", \"Type\", \"algorithm\", \"chr\", \"bkpt\", \"chrB\", \"bkptB\", \"RefStartPos\", \"RefEndPos\", \"Size\", \"Zygosity\", \"Confidence\", \"smapID\", \"linkSmapId\"]\n",
    "ctrldb_df.columns = ctrldb_headers\n",
    "ctrldb_df = ctrldb_df.astype({'RefStartPos':'float64', 'RefEndPos':'float64', 'chr':'int','Size':'float64'})\n",
    "# ctrldb_df = ctrldb_df.loc[~ctrldb_df['Type'].str.contains(\"trans\")]\n",
    "# ctrldb_df[\"has_gaps\"] = ctrldb_df.apply(lambda x: filter_nogaps(x[\"Type\"], x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "# ctrldb_df = ctrldb_df.loc[ctrldb_df[\"has_gaps\"] == False]\n",
    "# ctrldb_df = ctrldb_df.loc[ctrl_df['Sample'].str.contains(\"dle1\")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['trans', 'duplication_inverted', 'duplication',\n",
       "       'duplication_split', 'inv', 'ins', 'del'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 8.,  2.,  1., 23., 22., 15., 14., 10.,  9., 21., 20.,  7.,  5.,\n",
       "       11., 17.,  6., 24., 16.,  4., 13., 12., 19.])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df.loc[(ctrldb_df['Type'] == 'trans') & (ctrldb_df['chr'] == ctrldb_df['chrB']), 'chrB'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['unknown'], dtype=object)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df.loc[(ctrldb_df['Type'].str.contains('duplication')), 'Zygosity'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trans\n",
      "trans\n",
      "trans\n",
      "trans\n",
      "trans\n",
      "trans\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "      <th>%_controldb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7359682.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7950679.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7359682.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7950679.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7913502.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7398088.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7807846.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7513171.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7421</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7913502.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7398088.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample   Type            algorithm  chr       bkpt  chrB      bkptB  \\\n",
       "0  1_dle1_sop  trans  assembly_comparison    8  7359682.0   8.0  7950679.0   \n",
       "1  1_dle1_sop  trans  assembly_comparison    8  7359682.0   8.0  7950679.0   \n",
       "2  1_dle1_sop  trans  assembly_comparison    8  7913502.0   8.0  7398088.0   \n",
       "3  1_dle1_sop  trans  assembly_comparison    8  7807846.0   8.0  7513171.0   \n",
       "4  1_dle1_sop  trans  assembly_comparison    8  7913502.0   8.0  7398088.0   \n",
       "\n",
       "   RefStartPos  RefEndPos  Size      Zygosity  Confidence  smapID  linkSmapId  \\\n",
       "0          NaN        NaN   NaN  heterozygous        0.01    7408         NaN   \n",
       "1          NaN        NaN   NaN  heterozygous        0.01    7409         NaN   \n",
       "2          NaN        NaN   NaN  heterozygous        0.00    7417         NaN   \n",
       "3          NaN        NaN   NaN    homozygous        0.01    7421         NaN   \n",
       "4          NaN        NaN   NaN  heterozygous        0.00    7422         NaN   \n",
       "\n",
       "   %_controldb  \n",
       "0          0.0  \n",
       "1          0.0  \n",
       "2          0.0  \n",
       "3          0.0  \n",
       "4          0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "duplication_inverted\n",
      "duplication_inverted\n",
      "duplication_inverted\n",
      "duplication_inverted\n",
      "duplication_inverted\n",
      "duplication_inverted\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "      <th>%_controldb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37850</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_inverted</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13223159.0</td>\n",
       "      <td>13263012.0</td>\n",
       "      <td>39854.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.016760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37851</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_inverted</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16566162.0</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>50657.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.765363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37852</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_inverted</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16566162.0</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>50657.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>172</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.765363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37854</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_inverted</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16566162.0</td>\n",
       "      <td>16653300.0</td>\n",
       "      <td>87139.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.765363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37855</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_inverted</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16566162.0</td>\n",
       "      <td>16653300.0</td>\n",
       "      <td>87139.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.765363</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample                  Type            algorithm  chr  bkpt  chrB  \\\n",
       "37850  1_dle1_sop  duplication_inverted  assembly_comparison    1   NaN   NaN   \n",
       "37851  1_dle1_sop  duplication_inverted  assembly_comparison    1   NaN   NaN   \n",
       "37852  1_dle1_sop  duplication_inverted  assembly_comparison    1   NaN   NaN   \n",
       "37854  1_dle1_sop  duplication_inverted  assembly_comparison    1   NaN   NaN   \n",
       "37855  1_dle1_sop  duplication_inverted  assembly_comparison    1   NaN   NaN   \n",
       "\n",
       "       bkptB  RefStartPos   RefEndPos     Size Zygosity  Confidence  smapID  \\\n",
       "37850    NaN   13223159.0  13263012.0  39854.0  unknown        -1.0     169   \n",
       "37851    NaN   16566162.0  16616818.0  50657.0  unknown        -1.0     171   \n",
       "37852    NaN   16566162.0  16616818.0  50657.0  unknown        -1.0     172   \n",
       "37854    NaN   16566162.0  16653300.0  87139.0  unknown        -1.0     194   \n",
       "37855    NaN   16566162.0  16653300.0  87139.0  unknown        -1.0     195   \n",
       "\n",
       "       linkSmapId  %_controldb  \n",
       "37850         NaN    43.016760  \n",
       "37851         NaN    97.765363  \n",
       "37852         NaN    97.765363  \n",
       "37854         NaN    97.765363  \n",
       "37855         NaN    97.765363  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "duplication\n",
      "duplication\n",
      "duplication\n",
      "duplication\n",
      "duplication\n",
      "duplication\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "      <th>%_controldb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37853</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16566162.0</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>50657.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87.709497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37856</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79909385.0</td>\n",
       "      <td>79930051.0</td>\n",
       "      <td>20667.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>379</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.558659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37861</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37706418.0</td>\n",
       "      <td>37873078.0</td>\n",
       "      <td>166661.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1512</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.558659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37866</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20507397.0</td>\n",
       "      <td>20528202.0</td>\n",
       "      <td>20806.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.558659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37867</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>101306303.0</td>\n",
       "      <td>101336323.0</td>\n",
       "      <td>30021.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.558659</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample         Type            algorithm  chr  bkpt  chrB  bkptB  \\\n",
       "37853  1_dle1_sop  duplication  assembly_comparison    1   NaN   NaN    NaN   \n",
       "37856  1_dle1_sop  duplication  assembly_comparison    1   NaN   NaN    NaN   \n",
       "37861  1_dle1_sop  duplication  assembly_comparison    2   NaN   NaN    NaN   \n",
       "37866  1_dle1_sop  duplication  assembly_comparison    3   NaN   NaN    NaN   \n",
       "37867  1_dle1_sop  duplication  assembly_comparison    3   NaN   NaN    NaN   \n",
       "\n",
       "       RefStartPos    RefEndPos      Size Zygosity  Confidence  smapID  \\\n",
       "37853   16566162.0   16616818.0   50657.0  unknown        -1.0     185   \n",
       "37856   79909385.0   79930051.0   20667.0  unknown        -1.0     379   \n",
       "37861   37706418.0   37873078.0  166661.0  unknown        -1.0    1512   \n",
       "37866   20507397.0   20528202.0   20806.0  unknown        -1.0    2700   \n",
       "37867  101306303.0  101336323.0   30021.0  unknown        -1.0    3130   \n",
       "\n",
       "       linkSmapId  %_controldb  \n",
       "37853         NaN    87.709497  \n",
       "37856         NaN     0.558659  \n",
       "37861         NaN     0.558659  \n",
       "37866         NaN     0.558659  \n",
       "37867         NaN     0.558659  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "duplication_split\n",
      "duplication_split\n",
      "duplication_split\n",
      "duplication_split\n",
      "duplication_split\n",
      "duplication_split\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "      <th>%_controldb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37859</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_split</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30104624.0</td>\n",
       "      <td>31133651.0</td>\n",
       "      <td>1029028.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>663</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.558659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37862</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_split</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87161283.0</td>\n",
       "      <td>87737384.0</td>\n",
       "      <td>576102.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.675978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37883</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_split</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43058466.0</td>\n",
       "      <td>43090233.0</td>\n",
       "      <td>31768.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>8406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.821229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37884</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_split</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43058466.0</td>\n",
       "      <td>43090233.0</td>\n",
       "      <td>31768.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>8408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.821229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37885</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>duplication_split</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40878043.0</td>\n",
       "      <td>41912458.0</td>\n",
       "      <td>1034416.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>8411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.675978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample               Type            algorithm  chr  bkpt  chrB  \\\n",
       "37859  1_dle1_sop  duplication_split  assembly_comparison    1   NaN   NaN   \n",
       "37862  1_dle1_sop  duplication_split  assembly_comparison    2   NaN   NaN   \n",
       "37883  1_dle1_sop  duplication_split  assembly_comparison    9   NaN   NaN   \n",
       "37884  1_dle1_sop  duplication_split  assembly_comparison    9   NaN   NaN   \n",
       "37885  1_dle1_sop  duplication_split  assembly_comparison    9   NaN   NaN   \n",
       "\n",
       "       bkptB  RefStartPos   RefEndPos       Size Zygosity  Confidence  smapID  \\\n",
       "37859    NaN   30104624.0  31133651.0  1029028.0  unknown        -1.0     663   \n",
       "37862    NaN   87161283.0  87737384.0   576102.0  unknown        -1.0    1849   \n",
       "37883    NaN   43058466.0  43090233.0    31768.0  unknown        -1.0    8406   \n",
       "37884    NaN   43058466.0  43090233.0    31768.0  unknown        -1.0    8408   \n",
       "37885    NaN   40878043.0  41912458.0  1034416.0  unknown        -1.0    8411   \n",
       "\n",
       "       linkSmapId  %_controldb  \n",
       "37859         NaN     0.558659  \n",
       "37862         NaN     1.675978  \n",
       "37883         NaN     7.821229  \n",
       "37884         NaN     7.821229  \n",
       "37885         NaN     1.675978  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "      <th>%_controldb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>72899</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>inv</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>16883281.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>173</td>\n",
       "      <td>174.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72900</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>inv</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16616818.0</td>\n",
       "      <td>16859181.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>186</td>\n",
       "      <td>187.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72901</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>inv</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16653300.0</td>\n",
       "      <td>16859181.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>0.02</td>\n",
       "      <td>192</td>\n",
       "      <td>193.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72902</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>inv</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16653300.0</td>\n",
       "      <td>16653300.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>196</td>\n",
       "      <td>197.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72903</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>inv</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16653300.0</td>\n",
       "      <td>16859181.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>199</td>\n",
       "      <td>200.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample Type            algorithm  chr  bkpt  chrB  bkptB  \\\n",
       "72899  1_dle1_sop  inv  assembly_comparison    1   NaN   NaN    NaN   \n",
       "72900  1_dle1_sop  inv  assembly_comparison    1   NaN   NaN    NaN   \n",
       "72901  1_dle1_sop  inv  assembly_comparison    1   NaN   NaN    NaN   \n",
       "72902  1_dle1_sop  inv  assembly_comparison    1   NaN   NaN    NaN   \n",
       "72903  1_dle1_sop  inv  assembly_comparison    1   NaN   NaN    NaN   \n",
       "\n",
       "       RefStartPos   RefEndPos  Size      Zygosity  Confidence  smapID  \\\n",
       "72899   16616818.0  16883281.0  -1.0  heterozygous        0.00     173   \n",
       "72900   16616818.0  16859181.0  -1.0  heterozygous        0.00     186   \n",
       "72901   16653300.0  16859181.0  -1.0    homozygous        0.02     192   \n",
       "72902   16653300.0  16653300.0  -1.0  heterozygous        0.00     196   \n",
       "72903   16653300.0  16859181.0  -1.0    homozygous        0.00     199   \n",
       "\n",
       "       linkSmapId  %_controldb  \n",
       "72899       174.0        100.0  \n",
       "72900       187.0        100.0  \n",
       "72901       193.0        100.0  \n",
       "72902       197.0        100.0  \n",
       "72903       200.0        100.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ins\n",
      "ins\n",
      "ins\n",
      "ins\n",
      "ins\n",
      "ins\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "      <th>%_controldb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124814</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>ins</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>365535.0</td>\n",
       "      <td>387230.0</td>\n",
       "      <td>10175.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.99</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.407821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124816</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>ins</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>663861.0</td>\n",
       "      <td>672078.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.82</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.670391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124817</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>ins</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>710358.0</td>\n",
       "      <td>711832.0</td>\n",
       "      <td>336.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86.033520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124818</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>ins</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>788500.0</td>\n",
       "      <td>803475.0</td>\n",
       "      <td>2436.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.99</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124820</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>ins</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>858481.0</td>\n",
       "      <td>882186.0</td>\n",
       "      <td>2208.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.99</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.530726</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Sample Type            algorithm  chr  bkpt  chrB  bkptB  \\\n",
       "124814  1_dle1_sop  ins  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124816  1_dle1_sop  ins  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124817  1_dle1_sop  ins  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124818  1_dle1_sop  ins  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124820  1_dle1_sop  ins  assembly_comparison    1   NaN   NaN    NaN   \n",
       "\n",
       "        RefStartPos  RefEndPos     Size      Zygosity  Confidence  smapID  \\\n",
       "124814     365535.0   387230.0  10175.0  heterozygous        0.99       3   \n",
       "124816     663861.0   672078.0    536.0  heterozygous        0.82       5   \n",
       "124817     710358.0   711832.0    336.0  heterozygous       -1.00       6   \n",
       "124818     788500.0   803475.0   2436.0  heterozygous        0.99       7   \n",
       "124820     858481.0   882186.0   2208.0  heterozygous        0.99       9   \n",
       "\n",
       "        linkSmapId  %_controldb  \n",
       "124814         NaN    13.407821  \n",
       "124816         NaN    20.670391  \n",
       "124817         NaN    86.033520  \n",
       "124818         NaN   100.000000  \n",
       "124820         NaN    95.530726  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "del\n",
      "del\n",
      "del\n",
      "del\n",
      "del\n",
      "del\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "      <th>%_controldb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124815</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>del</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>588217.0</td>\n",
       "      <td>573534.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.99</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124819</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>del</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>817708.0</td>\n",
       "      <td>825653.0</td>\n",
       "      <td>1119.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.99</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.558659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124821</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>del</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>896304.0</td>\n",
       "      <td>912049.0</td>\n",
       "      <td>377.0</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.754190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124822</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>del</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>588217.0</td>\n",
       "      <td>573520.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.99</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124826</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>del</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>817708.0</td>\n",
       "      <td>825653.0</td>\n",
       "      <td>1119.0</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.99</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.558659</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Sample Type            algorithm  chr  bkpt  chrB  bkptB  \\\n",
       "124815  1_dle1_sop  del  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124819  1_dle1_sop  del  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124821  1_dle1_sop  del  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124822  1_dle1_sop  del  assembly_comparison    1   NaN   NaN    NaN   \n",
       "124826  1_dle1_sop  del  assembly_comparison    1   NaN   NaN    NaN   \n",
       "\n",
       "        RefStartPos  RefEndPos      Size      Zygosity  Confidence  smapID  \\\n",
       "124815          0.0   588217.0  573534.0  heterozygous        0.99       4   \n",
       "124819     817708.0   825653.0    1119.0  heterozygous        0.99       8   \n",
       "124821     896304.0   912049.0     377.0    homozygous       -1.00      10   \n",
       "124822          0.0   588217.0  573520.0  heterozygous        0.99      11   \n",
       "124826     817708.0   825653.0    1119.0  heterozygous        0.99      15   \n",
       "\n",
       "        linkSmapId  %_controldb  \n",
       "124815         NaN     0.000000  \n",
       "124819         NaN     0.558659  \n",
       "124821         NaN    35.754190  \n",
       "124822         NaN     0.000000  \n",
       "124826         NaN     0.558659  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for type in ctrldb_df['Type'].unique():\n",
    "    df = ctrldb_df.loc[ctrldb_df['Type'] == type].head()\n",
    "    df['%_controldb'] = df[['Type', 'chr', 'chrB', 'RefStartPos', 'RefEndPos', 'Size']].apply(lambda x: calculate_percent_presence_controldb(x[0], x[1], x[2], x[3], x[4], x[5]), axis=1)\n",
    "    print(type), display(df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Type</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>chr</th>\n",
       "      <th>bkpt</th>\n",
       "      <th>chrB</th>\n",
       "      <th>bkptB</th>\n",
       "      <th>RefStartPos</th>\n",
       "      <th>RefEndPos</th>\n",
       "      <th>Size</th>\n",
       "      <th>Zygosity</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>smapID</th>\n",
       "      <th>linkSmapId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7359682.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7950679.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7408</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7359682.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7950679.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7409</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7913502.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7398088.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7417</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7807846.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7513171.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7421</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_dle1_sop</td>\n",
       "      <td>trans</td>\n",
       "      <td>assembly_comparison</td>\n",
       "      <td>8</td>\n",
       "      <td>7913502.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7398088.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7422</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample   Type            algorithm  chr       bkpt  chrB      bkptB  \\\n",
       "0  1_dle1_sop  trans  assembly_comparison    8  7359682.0   8.0  7950679.0   \n",
       "1  1_dle1_sop  trans  assembly_comparison    8  7359682.0   8.0  7950679.0   \n",
       "2  1_dle1_sop  trans  assembly_comparison    8  7913502.0   8.0  7398088.0   \n",
       "3  1_dle1_sop  trans  assembly_comparison    8  7807846.0   8.0  7513171.0   \n",
       "4  1_dle1_sop  trans  assembly_comparison    8  7913502.0   8.0  7398088.0   \n",
       "\n",
       "   RefStartPos  RefEndPos  Size      Zygosity  Confidence  smapID  linkSmapId  \n",
       "0          NaN        NaN   NaN  heterozygous        0.01    7408         NaN  \n",
       "1          NaN        NaN   NaN  heterozygous        0.01    7409         NaN  \n",
       "2          NaN        NaN   NaN  heterozygous        0.00    7417         NaN  \n",
       "3          NaN        NaN   NaN    homozygous        0.01    7421         NaN  \n",
       "4          NaN        NaN   NaN  heterozygous        0.00    7422         NaN  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrldb_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starting with 3428999 total SVs\n",
      "2822201 SVs from dle1 labeling\n",
      "2787424 SVs after translocations removed\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\percent_presence_in_controldb.ipynb Cell 41\u001b[0m in \u001b[0;36m2\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m nogaps_df \u001b[39m=\u001b[39m notrans_df\u001b[39m.\u001b[39mloc[notrans_df[\u001b[39m\"\u001b[39m\u001b[39mhas_gaps\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m \u001b[39mFalse\u001b[39;00m]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m out_file(nogaps_df, \u001b[39m\"\u001b[39m\u001b[39mcontrol_nogaps_out\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m nogaps_df[\u001b[39m'\u001b[39m\u001b[39mPresent_in_\u001b[39m\u001b[39m%\u001b[39m\u001b[39m_of_BNG_control_samples_with_the_same_enzyme\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m nogaps_df[[\u001b[39m'\u001b[39;49m\u001b[39mType\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mchr\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mchrB\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mRefStartPos\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mRefEndPos\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mSize\u001b[39;49m\u001b[39m'\u001b[39;49m]]\u001b[39m.\u001b[39;49mapply(\u001b[39mlambda\u001b[39;49;00m x: calculate_percent_presence_controldb(x[\u001b[39m0\u001b[39;49m], x[\u001b[39m1\u001b[39;49m], x[\u001b[39m2\u001b[39;49m], x[\u001b[39m3\u001b[39;49m], x[\u001b[39m4\u001b[39;49m], x[\u001b[39m5\u001b[39;49m]), axis\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m rare_df \u001b[39m=\u001b[39m nogaps_df\u001b[39m.\u001b[39mloc[nogaps_df[\u001b[39m'\u001b[39m\u001b[39mPresent_in_\u001b[39m\u001b[39m%\u001b[39m\u001b[39m_of_BNG_control_samples_with_the_same_enzyme\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1.0\u001b[39m]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m out_file(rare_df, \u001b[39m\"\u001b[39m\u001b[39mcontrol_rare_out\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\frame.py:8848\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[1;34m(self, func, axis, raw, result_type, args, **kwargs)\u001b[0m\n\u001b[0;32m   8837\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpandas\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcore\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mapply\u001b[39;00m \u001b[39mimport\u001b[39;00m frame_apply\n\u001b[0;32m   8839\u001b[0m op \u001b[39m=\u001b[39m frame_apply(\n\u001b[0;32m   8840\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m   8841\u001b[0m     func\u001b[39m=\u001b[39mfunc,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   8846\u001b[0m     kwargs\u001b[39m=\u001b[39mkwargs,\n\u001b[0;32m   8847\u001b[0m )\n\u001b[1;32m-> 8848\u001b[0m \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mapply()\u001b[39m.\u001b[39m__finalize__(\u001b[39mself\u001b[39m, method\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mapply\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\apply.py:733\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    730\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw:\n\u001b[0;32m    731\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapply_raw()\n\u001b[1;32m--> 733\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_standard()\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\apply.py:857\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    856\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mapply_standard\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m--> 857\u001b[0m     results, res_index \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_series_generator()\n\u001b[0;32m    859\u001b[0m     \u001b[39m# wrap results\u001b[39;00m\n\u001b[0;32m    860\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\apply.py:873\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    870\u001b[0m \u001b[39mwith\u001b[39;00m option_context(\u001b[39m\"\u001b[39m\u001b[39mmode.chained_assignment\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):\n\u001b[0;32m    871\u001b[0m     \u001b[39mfor\u001b[39;00m i, v \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(series_gen):\n\u001b[0;32m    872\u001b[0m         \u001b[39m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[1;32m--> 873\u001b[0m         results[i] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mf(v)\n\u001b[0;32m    874\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[0;32m    875\u001b[0m             \u001b[39m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[0;32m    876\u001b[0m             \u001b[39m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[0;32m    877\u001b[0m             results[i] \u001b[39m=\u001b[39m results[i]\u001b[39m.\u001b[39mcopy(deep\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n",
      "\u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\percent_presence_in_controldb.ipynb Cell 41\u001b[0m in \u001b[0;36m2\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m nogaps_df \u001b[39m=\u001b[39m notrans_df\u001b[39m.\u001b[39mloc[notrans_df[\u001b[39m\"\u001b[39m\u001b[39mhas_gaps\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m \u001b[39mFalse\u001b[39;00m]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m out_file(nogaps_df, \u001b[39m\"\u001b[39m\u001b[39mcontrol_nogaps_out\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m nogaps_df[\u001b[39m'\u001b[39m\u001b[39mPresent_in_\u001b[39m\u001b[39m%\u001b[39m\u001b[39m_of_BNG_control_samples_with_the_same_enzyme\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m nogaps_df[[\u001b[39m'\u001b[39m\u001b[39mType\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mchr\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mchrB\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mRefStartPos\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mRefEndPos\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mSize\u001b[39m\u001b[39m'\u001b[39m]]\u001b[39m.\u001b[39mapply(\u001b[39mlambda\u001b[39;00m x: calculate_percent_presence_controldb(x[\u001b[39m0\u001b[39;49m], x[\u001b[39m1\u001b[39;49m], x[\u001b[39m2\u001b[39;49m], x[\u001b[39m3\u001b[39;49m], x[\u001b[39m4\u001b[39;49m], x[\u001b[39m5\u001b[39;49m]), axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m rare_df \u001b[39m=\u001b[39m nogaps_df\u001b[39m.\u001b[39mloc[nogaps_df[\u001b[39m'\u001b[39m\u001b[39mPresent_in_\u001b[39m\u001b[39m%\u001b[39m\u001b[39m_of_BNG_control_samples_with_the_same_enzyme\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1.0\u001b[39m]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m out_file(rare_df, \u001b[39m\"\u001b[39m\u001b[39mcontrol_rare_out\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\percent_presence_in_controldb.ipynb Cell 41\u001b[0m in \u001b[0;36m6\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=61'>62</a>\u001b[0m \u001b[39melif\u001b[39;00m Type \u001b[39m==\u001b[39m \u001b[39m'\u001b[39m\u001b[39mtrans\u001b[39m\u001b[39m'\u001b[39m:  \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=62'>63</a>\u001b[0m     posWindow, reciprocalSize \u001b[39m=\u001b[39m translocation_position_overlap, \u001b[39m0\u001b[39m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=64'>65</a>\u001b[0m match_chr \u001b[39m=\u001b[39m bed_df\u001b[39m.\u001b[39mloc[(bed_df[\u001b[39m\"\u001b[39m\u001b[39mchr\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m chrA) \u001b[39m&\u001b[39m (bed_df[\u001b[39m'\u001b[39;49m\u001b[39mType\u001b[39;49m\u001b[39m'\u001b[39;49m] \u001b[39m==\u001b[39;49m Type)]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=66'>67</a>\u001b[0m bed_start \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39marray(match_chr[\u001b[39m\"\u001b[39m\u001b[39mRefStartPos\u001b[39m\u001b[39m\"\u001b[39m])\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X54sZmlsZQ%3D%3D?line=67'>68</a>\u001b[0m bed_end \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39marray(match_chr[\u001b[39m\"\u001b[39m\u001b[39mRefEndPos\u001b[39m\u001b[39m\"\u001b[39m])\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\ops\\common.py:70\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     66\u001b[0m             \u001b[39mreturn\u001b[39;00m \u001b[39mNotImplemented\u001b[39m\n\u001b[0;32m     68\u001b[0m other \u001b[39m=\u001b[39m item_from_zerodim(other)\n\u001b[1;32m---> 70\u001b[0m \u001b[39mreturn\u001b[39;00m method(\u001b[39mself\u001b[39;49m, other)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\arraylike.py:40\u001b[0m, in \u001b[0;36mOpsMixin.__eq__\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     38\u001b[0m \u001b[39m@unpack_zerodim_and_defer\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m__eq__\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     39\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__eq__\u001b[39m(\u001b[39mself\u001b[39m, other):\n\u001b[1;32m---> 40\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_cmp_method(other, operator\u001b[39m.\u001b[39;49meq)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\series.py:5623\u001b[0m, in \u001b[0;36mSeries._cmp_method\u001b[1;34m(self, other, op)\u001b[0m\n\u001b[0;32m   5620\u001b[0m rvalues \u001b[39m=\u001b[39m extract_array(other, extract_numpy\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m, extract_range\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m   5622\u001b[0m \u001b[39mwith\u001b[39;00m np\u001b[39m.\u001b[39merrstate(\u001b[39mall\u001b[39m\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mignore\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m-> 5623\u001b[0m     res_values \u001b[39m=\u001b[39m ops\u001b[39m.\u001b[39;49mcomparison_op(lvalues, rvalues, op)\n\u001b[0;32m   5625\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_construct_result(res_values, name\u001b[39m=\u001b[39mres_name)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py:283\u001b[0m, in \u001b[0;36mcomparison_op\u001b[1;34m(left, right, op)\u001b[0m\n\u001b[0;32m    280\u001b[0m     \u001b[39mreturn\u001b[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001b[0;32m    282\u001b[0m \u001b[39melif\u001b[39;00m is_object_dtype(lvalues\u001b[39m.\u001b[39mdtype) \u001b[39mor\u001b[39;00m \u001b[39misinstance\u001b[39m(rvalues, \u001b[39mstr\u001b[39m):\n\u001b[1;32m--> 283\u001b[0m     res_values \u001b[39m=\u001b[39m comp_method_OBJECT_ARRAY(op, lvalues, rvalues)\n\u001b[0;32m    285\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    286\u001b[0m     res_values \u001b[39m=\u001b[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py:74\u001b[0m, in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[1;34m(op, x, y)\u001b[0m\n\u001b[0;32m     72\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m     73\u001b[0m     result \u001b[39m=\u001b[39m libops\u001b[39m.\u001b[39mscalar_compare(x\u001b[39m.\u001b[39mravel(), y, op)\n\u001b[1;32m---> 74\u001b[0m \u001b[39mreturn\u001b[39;00m result\u001b[39m.\u001b[39;49mreshape(x\u001b[39m.\u001b[39;49mshape)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Initialize dictionary to keepo track of filtering\n",
    "ctrldb_filtering = {}\n",
    "\n",
    "# Main code to clean up hg38 controlsdb .smap file\n",
    "print(f\"starting with {len(ctrldb_df)} total SVs\")\n",
    "\n",
    "# Keep only SVs detected with 'dle1' enzyme\n",
    "dle_df = ctrldb_df.loc[ctrldb_df[\"Sample\"].str.contains(\"dle1\")]\n",
    "print(f\"{len(dle_df)} SVs from dle1 labeling\")\n",
    "\n",
    "# remove translocations\n",
    "notrans_df = dle_df.loc[~dle_df['Type'].str.contains(\"trans\")]\n",
    "print(f\"{len(notrans_df)} SVs after translocations removed\")\n",
    "\n",
    "# # filter controlsdb by score filter\n",
    "# sig_df = filter_sig_svs(notrans_df)\n",
    "# print(f\"{len(sig_df)} SVs are significant, with confidence scores above Access recommended threshold\")\n",
    "\n",
    "# out_file(sig_df, \"control_sig_out\")\n",
    "\n",
    "# remove indel_ngaps\n",
    "# True if overlaps gaps, false if not, remove Falses\n",
    "notrans_df[\"has_gaps\"] = notrans_df.apply(lambda x: filter_nogaps(x[\"Type\"], x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "nogaps_df = notrans_df.loc[notrans_df[\"has_gaps\"] == False]\n",
    "\n",
    "out_file(nogaps_df, \"control_nogaps_out\")\n",
    "\n",
    "nogaps_df['Present_in_%_of_BNG_control_samples_with_the_same_enzyme'] = nogaps_df[['Type', 'chr', 'chrB', 'RefStartPos', 'RefEndPos', 'Size']].apply(lambda x: calculate_percent_presence_controldb(x[0], x[1], x[2], x[3], x[4], x[5]), axis=1)\n",
    "rare_df = nogaps_df.loc[nogaps_df['Present_in_%_of_BNG_control_samples_with_the_same_enzyme'] <= 1.0]\n",
    "out_file(rare_df, \"control_rare_out\")\n",
    "\n",
    "# filter controlsdb by score filter\n",
    "sig_df = filter_sig_svs(rare_df)\n",
    "print(f\"{len(sig_df)} SVs are significant, with confidence scores above Access recommended threshold\")\n",
    "\n",
    "out_file(sig_df, \"control_sig_out\")\n",
    "\n",
    "num_true = len(sig_df[sig_df[\"has_gaps\"] == True])\n",
    "num_false = len(sig_df[sig_df[\"has_gaps\"] == False])\n",
    "print(f\"{num_true} has gaps, and {num_false} has no gaps, for a total of {num_true + num_false} SVs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file(nogaps_df, \"control_nogaps_percentP{resence_out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\percent_presence_in_controldb.ipynb Cell 37\u001b[0m in \u001b[0;36m6\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m nogaps_df \u001b[39m=\u001b[39m nogaps_df\u001b[39m.\u001b[39mastype({\u001b[39m'\u001b[39m\u001b[39mRefStartPos\u001b[39m\u001b[39m'\u001b[39m:\u001b[39m'\u001b[39m\u001b[39mfloat64\u001b[39m\u001b[39m'\u001b[39m,\u001b[39m'\u001b[39m\u001b[39mRefEndPos\u001b[39m\u001b[39m'\u001b[39m:\u001b[39m'\u001b[39m\u001b[39mfloat64\u001b[39m\u001b[39m'\u001b[39m,\u001b[39m'\u001b[39m\u001b[39mchr\u001b[39m\u001b[39m'\u001b[39m:\u001b[39m'\u001b[39m\u001b[39mint\u001b[39m\u001b[39m'\u001b[39m})\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m nogaps_df\u001b[39m.\u001b[39mdtypes\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m nogaps_df[\u001b[39m\"\u001b[39m\u001b[39mOverlapGenes\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m nogaps_df\u001b[39m.\u001b[39;49mapply(\u001b[39mlambda\u001b[39;49;00m x: filter_genes(bed_df, x[\u001b[39m\"\u001b[39;49m\u001b[39mchr\u001b[39;49m\u001b[39m\"\u001b[39;49m], x[\u001b[39m\"\u001b[39;49m\u001b[39mRefStartPos\u001b[39;49m\u001b[39m\"\u001b[39;49m], x[\u001b[39m\"\u001b[39;49m\u001b[39mRefEndPos\u001b[39;49m\u001b[39m\"\u001b[39;49m]), axis\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m out_file(nogaps_df, \u001b[39m\"\u001b[39m\u001b[39mcontrol_gene_nogaps_sig_out\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m no_overlaps \u001b[39m=\u001b[39m \u001b[39mlen\u001b[39m(nogaps_df\u001b[39m.\u001b[39mloc[nogaps_df[\u001b[39m\"\u001b[39m\u001b[39mOverlapGenes\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m-\u001b[39m\u001b[39m\"\u001b[39m])\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\frame.py:8848\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[1;34m(self, func, axis, raw, result_type, args, **kwargs)\u001b[0m\n\u001b[0;32m   8837\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpandas\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcore\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mapply\u001b[39;00m \u001b[39mimport\u001b[39;00m frame_apply\n\u001b[0;32m   8839\u001b[0m op \u001b[39m=\u001b[39m frame_apply(\n\u001b[0;32m   8840\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m   8841\u001b[0m     func\u001b[39m=\u001b[39mfunc,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   8846\u001b[0m     kwargs\u001b[39m=\u001b[39mkwargs,\n\u001b[0;32m   8847\u001b[0m )\n\u001b[1;32m-> 8848\u001b[0m \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mapply()\u001b[39m.\u001b[39m__finalize__(\u001b[39mself\u001b[39m, method\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mapply\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\apply.py:733\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    730\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw:\n\u001b[0;32m    731\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapply_raw()\n\u001b[1;32m--> 733\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_standard()\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\apply.py:857\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    856\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mapply_standard\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m--> 857\u001b[0m     results, res_index \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_series_generator()\n\u001b[0;32m    859\u001b[0m     \u001b[39m# wrap results\u001b[39;00m\n\u001b[0;32m    860\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\apply.py:873\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    870\u001b[0m \u001b[39mwith\u001b[39;00m option_context(\u001b[39m\"\u001b[39m\u001b[39mmode.chained_assignment\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):\n\u001b[0;32m    871\u001b[0m     \u001b[39mfor\u001b[39;00m i, v \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(series_gen):\n\u001b[0;32m    872\u001b[0m         \u001b[39m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[1;32m--> 873\u001b[0m         results[i] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mf(v)\n\u001b[0;32m    874\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[0;32m    875\u001b[0m             \u001b[39m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[0;32m    876\u001b[0m             \u001b[39m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[0;32m    877\u001b[0m             results[i] \u001b[39m=\u001b[39m results[i]\u001b[39m.\u001b[39mcopy(deep\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n",
      "\u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\percent_presence_in_controldb.ipynb Cell 37\u001b[0m in \u001b[0;36m6\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m nogaps_df \u001b[39m=\u001b[39m nogaps_df\u001b[39m.\u001b[39mastype({\u001b[39m'\u001b[39m\u001b[39mRefStartPos\u001b[39m\u001b[39m'\u001b[39m:\u001b[39m'\u001b[39m\u001b[39mfloat64\u001b[39m\u001b[39m'\u001b[39m,\u001b[39m'\u001b[39m\u001b[39mRefEndPos\u001b[39m\u001b[39m'\u001b[39m:\u001b[39m'\u001b[39m\u001b[39mfloat64\u001b[39m\u001b[39m'\u001b[39m,\u001b[39m'\u001b[39m\u001b[39mchr\u001b[39m\u001b[39m'\u001b[39m:\u001b[39m'\u001b[39m\u001b[39mint\u001b[39m\u001b[39m'\u001b[39m})\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m nogaps_df\u001b[39m.\u001b[39mdtypes\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m nogaps_df[\u001b[39m\"\u001b[39m\u001b[39mOverlapGenes\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m nogaps_df\u001b[39m.\u001b[39mapply(\u001b[39mlambda\u001b[39;00m x: filter_genes(bed_df, x[\u001b[39m\"\u001b[39;49m\u001b[39mchr\u001b[39;49m\u001b[39m\"\u001b[39;49m], x[\u001b[39m\"\u001b[39;49m\u001b[39mRefStartPos\u001b[39;49m\u001b[39m\"\u001b[39;49m], x[\u001b[39m\"\u001b[39;49m\u001b[39mRefEndPos\u001b[39;49m\u001b[39m\"\u001b[39;49m]), axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m out_file(nogaps_df, \u001b[39m\"\u001b[39m\u001b[39mcontrol_gene_nogaps_sig_out\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m no_overlaps \u001b[39m=\u001b[39m \u001b[39mlen\u001b[39m(nogaps_df\u001b[39m.\u001b[39mloc[nogaps_df[\u001b[39m\"\u001b[39m\u001b[39mOverlapGenes\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m-\u001b[39m\u001b[39m\"\u001b[39m])\n",
      "\u001b[1;32mc:\\Users\\sshukor\\OneDrive - Bionano Genomics\\Documents\\Enrichment Project\\clustering_optimization_enrichment\\percent_presence_in_controldb.ipynb Cell 37\u001b[0m in \u001b[0;36m4\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mfilter_genes\u001b[39m(bed_df, \u001b[39mchr\u001b[39m, start, end):\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m     match_chr \u001b[39m=\u001b[39m bed_df\u001b[39m.\u001b[39mloc[(bed_df[\u001b[39m\"\u001b[39;49m\u001b[39mchr\u001b[39;49m\u001b[39m\"\u001b[39;49m] \u001b[39m==\u001b[39;49m \u001b[39mchr\u001b[39;49m)]\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m     \u001b[39m# display(match_chr)\u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m \n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m     \u001b[39m# transform matched chr rows into numpy (faster processing than directly panda slicing)\u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/sshukor/OneDrive%20-%20Bionano%20Genomics/Documents/Enrichment%20Project/clustering_optimization_enrichment/percent_presence_in_controldb.ipynb#X45sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     bed_genes \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39marray(match_chr[\u001b[39m\"\u001b[39m\u001b[39mGene\u001b[39m\u001b[39m\"\u001b[39m])\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\ops\\common.py:70\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     66\u001b[0m             \u001b[39mreturn\u001b[39;00m \u001b[39mNotImplemented\u001b[39m\n\u001b[0;32m     68\u001b[0m other \u001b[39m=\u001b[39m item_from_zerodim(other)\n\u001b[1;32m---> 70\u001b[0m \u001b[39mreturn\u001b[39;00m method(\u001b[39mself\u001b[39;49m, other)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\arraylike.py:40\u001b[0m, in \u001b[0;36mOpsMixin.__eq__\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     38\u001b[0m \u001b[39m@unpack_zerodim_and_defer\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m__eq__\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     39\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__eq__\u001b[39m(\u001b[39mself\u001b[39m, other):\n\u001b[1;32m---> 40\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_cmp_method(other, operator\u001b[39m.\u001b[39;49meq)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\series.py:5623\u001b[0m, in \u001b[0;36mSeries._cmp_method\u001b[1;34m(self, other, op)\u001b[0m\n\u001b[0;32m   5620\u001b[0m rvalues \u001b[39m=\u001b[39m extract_array(other, extract_numpy\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m, extract_range\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m   5622\u001b[0m \u001b[39mwith\u001b[39;00m np\u001b[39m.\u001b[39merrstate(\u001b[39mall\u001b[39m\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mignore\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m-> 5623\u001b[0m     res_values \u001b[39m=\u001b[39m ops\u001b[39m.\u001b[39;49mcomparison_op(lvalues, rvalues, op)\n\u001b[0;32m   5625\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_construct_result(res_values, name\u001b[39m=\u001b[39mres_name)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py:286\u001b[0m, in \u001b[0;36mcomparison_op\u001b[1;34m(left, right, op)\u001b[0m\n\u001b[0;32m    283\u001b[0m     res_values \u001b[39m=\u001b[39m comp_method_OBJECT_ARRAY(op, lvalues, rvalues)\n\u001b[0;32m    285\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 286\u001b[0m     res_values \u001b[39m=\u001b[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)\n\u001b[0;32m    288\u001b[0m \u001b[39mreturn\u001b[39;00m res_values\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py:163\u001b[0m, in \u001b[0;36m_na_arithmetic_op\u001b[1;34m(left, right, op, is_cmp)\u001b[0m\n\u001b[0;32m    160\u001b[0m     func \u001b[39m=\u001b[39m partial(expressions\u001b[39m.\u001b[39mevaluate, op)\n\u001b[0;32m    162\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 163\u001b[0m     result \u001b[39m=\u001b[39m func(left, right)\n\u001b[0;32m    164\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[0;32m    165\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m is_cmp \u001b[39mand\u001b[39;00m (is_object_dtype(left\u001b[39m.\u001b[39mdtype) \u001b[39mor\u001b[39;00m is_object_dtype(right)):\n\u001b[0;32m    166\u001b[0m         \u001b[39m# For object dtype, fallback to a masked operation (only operating\u001b[39;00m\n\u001b[0;32m    167\u001b[0m         \u001b[39m#  on the non-missing values)\u001b[39;00m\n\u001b[0;32m    168\u001b[0m         \u001b[39m# Don't do this for comparisons, as that will handle complex numbers\u001b[39;00m\n\u001b[0;32m    169\u001b[0m         \u001b[39m#  incorrectly, see GH#32047\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\computation\\expressions.py:239\u001b[0m, in \u001b[0;36mevaluate\u001b[1;34m(op, a, b, use_numexpr)\u001b[0m\n\u001b[0;32m    236\u001b[0m \u001b[39mif\u001b[39;00m op_str \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    237\u001b[0m     \u001b[39mif\u001b[39;00m use_numexpr:\n\u001b[0;32m    238\u001b[0m         \u001b[39m# error: \"None\" not callable\u001b[39;00m\n\u001b[1;32m--> 239\u001b[0m         \u001b[39mreturn\u001b[39;00m _evaluate(op, op_str, a, b)  \u001b[39m# type: ignore[misc]\u001b[39;00m\n\u001b[0;32m    240\u001b[0m \u001b[39mreturn\u001b[39;00m _evaluate_standard(op, op_str, a, b)\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\computation\\expressions.py:128\u001b[0m, in \u001b[0;36m_evaluate_numexpr\u001b[1;34m(op, op_str, a, b)\u001b[0m\n\u001b[0;32m    125\u001b[0m     _store_test_result(result \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m)\n\u001b[0;32m    127\u001b[0m \u001b[39mif\u001b[39;00m result \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m--> 128\u001b[0m     result \u001b[39m=\u001b[39m _evaluate_standard(op, op_str, a, b)\n\u001b[0;32m    130\u001b[0m \u001b[39mreturn\u001b[39;00m result\n",
      "File \u001b[1;32mc:\\Users\\sshukor\\Anaconda3\\envs\\cs109a_1\\lib\\site-packages\\pandas\\core\\computation\\expressions.py:69\u001b[0m, in \u001b[0;36m_evaluate_standard\u001b[1;34m(op, op_str, a, b)\u001b[0m\n\u001b[0;32m     67\u001b[0m \u001b[39mif\u001b[39;00m _TEST_MODE:\n\u001b[0;32m     68\u001b[0m     _store_test_result(\u001b[39mFalse\u001b[39;00m)\n\u001b[1;32m---> 69\u001b[0m \u001b[39mreturn\u001b[39;00m op(a, b)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# find overlap genes matching hg38knownCanonical.bed\n",
    "nogaps_df = in_file(\"control_nogaps_sig_out\")\n",
    "nogaps_df = nogaps_df.astype({'RefStartPos':'float64','RefEndPos':'float64','chr':'int'})\n",
    "nogaps_df.dtypes\n",
    "\n",
    "nogaps_df[\"OverlapGenes\"] = nogaps_df.apply(lambda x: filter_genes(bed_df, x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "out_file(nogaps_df, \"control_gene_nogaps_sig_out\")\n",
    "\n",
    "no_overlaps = len(nogaps_df.loc[nogaps_df[\"OverlapGenes\"] == \"-\"])\n",
    "overlaps = len(nogaps_df.loc[nogaps_df[\"OverlapGenes\"] != \"-\"])\n",
    "print(f\"From {no_overlaps + overlaps} cleaned SVs, {no_overlaps} SVs do not overlap any gene, while {overlaps} do.\")\n",
    "\n",
    "out_file(nogaps_df, \"control_cleaned_out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Load files and Clean Pseudogenes bed file\n",
    "# # headers: chrom\tchromStart\tchromEnd\tGene\tIndex\tstrand\tchromStart2\tchromEnd\tRGB\n",
    "# # bed_headers = [\"chr\",\"RefStartPos\",\"RefEndPos\",\"Gene\",\"Index\",\"Strand\",\"RefStartPos2\",\"RefEndPos2\",\"RGB\"]\n",
    "\n",
    "# # Load cleaned ctrl_db file\n",
    "# # nogaps_df = pd.read_csv(r\"Output\\output_04Jan2023\\control_gene_nogaps_sig_out.csv\", index_col=False)\n",
    "# nogaps_df = in_file('control_gene_nogaps_sig_out')\n",
    "# nogaps_df = nogaps_df.astype({'RefStartPos':'float64','RefEndPos':'float64','chr':'int'})\n",
    "# display(nogaps_df.head())\n",
    "\n",
    "# # Check if SVs overlap which genes\n",
    "# nogaps_df[\"OverlapPseudogenes\"] = nogaps_df.apply(lambda x: filter_genes(pseudo_df, x[\"chr\"], x[\"RefStartPos\"], x[\"RefEndPos\"]), axis=1)\n",
    "\n",
    "# # remove SVs that overlap only pseudogenes\n",
    "# overlap_pseudogenes_df = nogaps_df.loc[((nogaps_df[\"OverlapGenes\"].str.split(';') == nogaps_df[\"OverlapPseudogenes\"].str.split(';')) & (nogaps_df[\"OverlapGenes\"] != \"-\"))\n",
    "#                                     | ((nogaps_df[\"OverlapPseudogenes\"] != \"-\") & (nogaps_df[\"OverlapGenes\"] == \"-\"))]\n",
    "# out_file(overlap_pseudogenes_df, \"control_pseudogenes_out\")\n",
    "\n",
    "# # Exclude SVs that are only overlapped by pseudogenes\n",
    "# cleaned_df = nogaps_df.loc[~(((nogaps_df[\"OverlapGenes\"].str.split(';') == nogaps_df[\"OverlapPseudogenes\"].str.split(';')) & (nogaps_df[\"OverlapGenes\"] != \"-\")) \n",
    "#                             |((nogaps_df[\"OverlapGenes\"] == \"-\") & (nogaps_df[\"OverlapPseudogenes\"] != \"-\")))]\n",
    "# # # Exclude SVs that are only overlapped by pseudogenes\n",
    "# # cleaned_df = case_df.loc[~((case_df[\"OverlapGenes\"] == case_df[\"OverlapPseudogenes\"]) & (case_df[\"OverlapGenes\"] != \"-\") \n",
    "# #                         | ((case_df[\"OverlapGenes\"] == \"-\") & (case_df[\"OverlapPseudogenes\"] != \"-\")))]\n",
    "\n",
    "# print(f\"There are {len(nogaps_df)} SVs before cleaning, {len(overlap_pseudogenes_df)} SVs are found to overlap only pseudogenes, leaving {len(cleaned_df)} SVs overlapping no genes or knownCanonical\")\n",
    "\n",
    "# # Excluded pseudogenes from controlsdb file\n",
    "# out_file(cleaned_df, \"control_cleaned_out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load cleaned ctrl_db file\n",
    "# cleaned_df = pd.read_csv(r\"Output\\output_06Jan2023\\control_cleaned_out.csv\", index_col=False)\n",
    "# cleaned_df = in_file('control_cleaned_out')\n",
    "# cleaned_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['trans' 'duplication_inverted' 'duplication' 'duplication_split' 'inv'\n",
      " 'ins' 'del']\n",
      "['duplication_inverted' 'duplication' 'duplication_split' 'inv' 'ins'\n",
      " 'del']\n",
      "['duplication_inverted' 'duplication' 'duplication_split' 'inv' 'ins'\n",
      " 'del']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Total SVs': 3428999,\n",
       " 'DLE1 Labelled SVs': 2822201,\n",
       " 'No translocations': 2787424,\n",
       " 'Significant SVs': 1193403,\n",
       " 'No n-base gaps': 1156665}"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Append to dictionary\n",
    "ctrldb_filtering={}\n",
    "ctrldb_filtering[\"Total SVs\"] = len(ctrldb_df)\n",
    "ctrldb_filtering[\"DLE1 Labelled SVs\"] = len(dle_df)\n",
    "print(pd.unique(dle_df['Type']))\n",
    "ctrldb_filtering[\"No translocations\"] = len(notrans_df)\n",
    "ctrldb_filtering[\"Significant SVs\"] = len(sig_df)\n",
    "print(pd.unique(sig_df['Type']))\n",
    "ctrldb_filtering[\"No n-base gaps\"] = (len(nogaps_df))\n",
    "print(pd.unique(nogaps_df['Type']))\n",
    "# ctrldb_filtering[\"Removed Pseudogene-only Overlaps\"] = (len(cleaned_df))\n",
    "ctrldb_filtering"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('cs109a_1')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "efcd0302cabbf2b1644af678674dd9a1a01243194fbb7441d63a1ded11bd7cd9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
